{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA Expression\n",
    "\n",
    "With RNA sequencing we have the promise of a measurement that should help us understand the actions and reactions of a cell.\n",
    "\n",
    "One of the findings of the genomic analysis found less protein coding differences than originally expected.  RNA expression levels and feed back control look like the major differences in human evolution between humans and chimps.\n",
    " * 'there are significant differences in how genes are expressed and regulated'\n",
    " * '.. those differences are most marked in the brain'\n",
    " * https://science.sciencemag.org/content/292/5514/44\n",
    " * '.. rapid change, along the five million years of the human lineage, that was concentrated on these specific groups of genes.'\n",
    " * 'Nearly half of the genes that had been pushed to express themselves more in humans involved transcription factors--gene-encoded proteins that control the expression of other proteins'\n",
    " * https://www.scientificamerican.com/article/separation-of-man-and-ape/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Expression Level Differences can complicated\n",
    "\n",
    "What is a 'significant' difference in expression?\n",
    "\n",
    "### Small differences can have a large effect\n",
    "\n",
    "Classic blond hair vs brunette in europeans is caused by 20% decrease in mRNA expression.\n",
    " * Caused by a SNP 350 kb away from KITLG coding region\n",
    " * SNP alters transcription factor binding\n",
    " * only 20% change is less than most check for significance.\n",
    " * https://www.nature.com/articles/ng.3019 Guether et al. 2014\n",
    "\n",
    "### But not always\n",
    " * [mRNA] / [protein] ratio varies by several orders of magnitude across genes.\n",
    " * but is consistent across cells and tissues, for each gene\n",
    " * Quantifying gene expression: the importance of being subtle\n",
    " * https://www.embopress.org/doi/pdf/10.15252/msb.20167325\n",
    "\n",
    "![Variations in mRNA vs Protein levels](img/Gene_expression_subtle.png)\n",
    "\n",
    "### Population at large can have outliers\n",
    "Can be large differences in normal population.\n",
    " * The impact of rare variation on gene expression across tissues\n",
    "  * 58% of underexpression and 28% of overexpression outliers have nearby conserved rare variants compared to 8% of non-outliers\n",
    "  * bayesian improved calling\n",
    " * https://www.ncbi.nlm.nih.gov/pubmed/29022581"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Exploration - Nature of the beast\n",
    "\n",
    "Most uses of expression data will use a comparison of the gene expression pattern to known references.\n",
    "What would this involve.\n",
    "\n",
    "## Normal Tissue Expressions\n",
    "\n",
    "Genotype-Tissue Expression (GTEx) project is becoming a major resource for normal human tissue.\n",
    " * v8 to be released this month.\n",
    "\n",
    "The spread includes extreme values, most of the tissues are similar.\n",
    "\n",
    "Mean is ~ 16 - 17 TPM, but 50th percentile is 0 and 75th percentile is ~ 2.\n",
    "\n",
    "This indicates that the TPM is not a normal distribution, but has a small number of extreme values.  Almost half of the 56K gene set show no expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Unnamed: 0  Adipose Tissue  Adrenal Gland       Bladder          Blood  \\\n",
      "0      count    56202.000000   56202.000000  56202.000000   56202.000000   \n",
      "1       mean       16.335619      17.107780     16.663300      15.233666   \n",
      "2        std      398.650451     582.642470    351.055527    1113.018285   \n",
      "3        min        0.000000       0.000000      0.000000       0.000000   \n",
      "4        25%        0.000000       0.000000      0.000000       0.000000   \n",
      "5        50%        0.000000       0.000000      0.000000       0.000000   \n",
      "6        75%        1.897000       1.492375      2.668000       0.492100   \n",
      "7        max    38300.000000   60775.000000  34110.000000  246600.000000   \n",
      "\n",
      "   Blood Vessel         Brain        Breast  Cervix Uteri         Colon  ...  \\\n",
      "0  56202.000000  56202.000000  56202.000000  56202.000000  56202.000000  ...   \n",
      "1     16.720560     16.984681     16.138433     16.543685     16.154840  ...   \n",
      "2    247.194939    762.406001    375.471731    238.259867    514.382505  ...   \n",
      "3      0.000000      0.000000      0.000000      0.000000      0.000000  ...   \n",
      "4      0.000000      0.000000      0.000000      0.000000      0.000000  ...   \n",
      "5      0.000000      0.000000      0.028010      0.036405      0.019235  ...   \n",
      "6      1.727000      1.503875      2.360500      3.063750      2.117500  ...   \n",
      "7  20530.000000  88690.000000  34805.000000  24350.000000  53740.000000  ...   \n",
      "\n",
      "       Prostate  Salivary Gland          Skin  Small Intestine        Spleen  \\\n",
      "0  56202.000000    56202.000000  56202.000000     56202.000000  56202.000000   \n",
      "1     16.227906       14.989199     16.792584        15.809636     16.516674   \n",
      "2    410.755096      341.015132    350.311541       464.867976    287.035114   \n",
      "3      0.000000        0.000000      0.000000         0.000000      0.000000   \n",
      "4      0.000000        0.000000      0.000000         0.000000      0.000000   \n",
      "5      0.049132        0.034170      0.021895         0.040825      0.032045   \n",
      "6      3.038500        2.283750      2.244750         2.487750      2.704750   \n",
      "7  40380.000000    29230.000000  34240.000000     45570.000000  26200.000000   \n",
      "\n",
      "        Stomach        Testis       Thyroid        Uterus        Vagina  \n",
      "0  56202.000000  56202.000000  56202.000000  56202.000000  56202.000000  \n",
      "1     15.245102     17.319775     16.608929     16.584184     15.826178  \n",
      "2    506.661336    359.338580    343.872469    247.546841    246.663196  \n",
      "3      0.000000      0.000000      0.000000      0.000000      0.000000  \n",
      "4      0.000000      0.000000      0.000000      0.000000      0.000000  \n",
      "5      0.000000      0.521350      0.045683      0.016190      0.030570  \n",
      "6      1.625875      6.391500      3.350125      2.806750      2.784000  \n",
      "7  44110.000000  35450.000000  34950.000000  22430.000000  21010.000000  \n",
      "\n",
      "[8 rows x 31 columns]\n"
     ]
    }
   ],
   "source": [
    "import importlib\n",
    "import os, sys, glob\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pathlib\n",
    "script_dir = pathlib.Path().resolve()\n",
    "\n",
    "gm_fn = os.path.join(script_dir, 'tables', 'GTEX_tissue_medians_desc.tsv')\n",
    "gm = pd.read_csv(gm_fn, sep='\\t')\n",
    "print(gm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most oncogenes and tumour suppressors are among genes with higher expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Unnamed: 0  Adipose Tissue  Adrenal Gland      Bladder        Blood  \\\n",
      "0      count      779.000000     779.000000   779.000000   779.000000   \n",
      "1       mean       50.191434      33.618767    49.578746    50.097736   \n",
      "2        std      158.616941     101.872038   131.885045   360.652820   \n",
      "3        min        0.000000       0.000000     0.000000     0.000000   \n",
      "4        25%        4.507500       2.593250     5.602000     0.576500   \n",
      "5        50%       14.680000      10.005000    18.820000     5.052000   \n",
      "6        75%       33.765000      25.515000    41.905000    20.595000   \n",
      "7        max     1979.000000    1084.500000  1770.000000  8644.000000   \n",
      "\n",
      "   Blood Vessel       Brain       Breast  Cervix Uteri        Colon  ...  \\\n",
      "0    779.000000  779.000000   779.000000    779.000000   779.000000  ...   \n",
      "1     57.683957   18.923307    50.954532     63.084032    38.090100  ...   \n",
      "2    168.540744   53.215936   160.465959    195.045619   121.327983  ...   \n",
      "3      0.000000    0.000000     0.000000      0.000000     0.000000  ...   \n",
      "4      3.297500    1.831750     5.453750      6.047000     4.496250  ...   \n",
      "5     16.870000    6.026500    16.730000     21.100000    13.000000  ...   \n",
      "6     43.715000   14.645000    37.537500     49.905000    28.300000  ...   \n",
      "7   2477.000000  727.200000  1958.500000   2521.000000  1705.500000  ...   \n",
      "\n",
      "      Prostate  Salivary Gland         Skin  Small Intestine       Spleen  \\\n",
      "0   779.000000      779.000000   779.000000       779.000000   779.000000   \n",
      "1    43.470492       37.984779    49.551988        44.064486    65.347761   \n",
      "2   118.400366      112.105753   153.299167       163.859053   316.526407   \n",
      "3     0.000000        0.000000     0.000000         0.000000     0.000000   \n",
      "4     5.514250        4.577500     4.605000         4.874000     5.409000   \n",
      "5    16.155000       13.380000    15.140000        14.110000    17.190000   \n",
      "6    34.550000       29.090000    37.740000        30.645000    43.902500   \n",
      "7  1444.000000     1559.000000  1915.000000      2705.000000  6267.500000   \n",
      "\n",
      "      Stomach      Testis      Thyroid       Uterus       Vagina  \n",
      "0   779.00000  779.000000   779.000000   779.000000   779.000000  \n",
      "1    30.61774   33.494211    57.345001    61.375940    52.343328  \n",
      "2    96.48838   69.849530   164.552883   193.035895   160.712867  \n",
      "3     0.00000    0.000000     0.000000     0.000000     0.000000  \n",
      "4     3.17875    4.687500     5.665250     5.232000     5.878500  \n",
      "5     9.53200   15.700000    19.345000    20.620000    17.940000  \n",
      "6    22.10750   33.710000    41.197500    50.970000    40.790000  \n",
      "7  1396.00000  890.300000  1761.000000  2574.000000  2404.000000  \n",
      "\n",
      "[8 rows x 31 columns]\n"
     ]
    }
   ],
   "source": [
    "gm_fn = os.path.join(script_dir, 'tables', 'GTEX_tissue_medians_onco_ts.tsv')\n",
    "gm = pd.read_csv(gm_fn, sep='\\t')\n",
    "print(gm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Natural Distribution is log normal\n",
    "\n",
    "This means the log of the values shows a normal distribution.\n",
    "\n",
    "### PTEN shows the normal distribution after log transform\n",
    "\n",
    "![](img/PTEN_raw.png)\n",
    "![](img/PTEN_log2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What's happening with TP53\n",
    "\n",
    "What does the camel like curve mean?\n",
    "\n",
    "![](img/TP53_raw.png)\n",
    "![](img/TP53_log2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What's happening with TP53 - Different Expression by tissue\n",
    "\n",
    "![](img/TP53_by_tissue.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experimental Variantions\n",
    "\n",
    "\n",
    "### Units RPKM, FPKM, TPM\n",
    "\n",
    "RPKM (single-end data) - Reads per\n",
    "FPKM (paired-end data) - Fragments per\n",
    "TPM (transcripts per million)\n",
    "\n",
    "RPKM and FPKM are normalized within a sample\n",
    "\n",
    "TPM is more normalized for comparing between samples\n",
    "\n",
    "\n",
    "Convert to TPM simply by dividing each RPKM value by the sum of the RPKM values for all genes (or transcripts) and multiplying by one million.\n",
    "\n",
    "TPM = (mean transcript length in kilobases) x RPKM \n",
    "where \"mean transcript length\" is the expression-weighted mean of the lengths of all isoforms.  Because the mean transcript length can change from sample to sample, we have generally recommended the use of TPM instead of RPKM.\n",
    "\n",
    "From the Google groups discussion: (https://groups.google.com/forum/#!topic/rsem-users/W9RQrZIOzA4)\n",
    "\n",
    "\n",
    "If you have RPKM (single-end data) or FPKM (paired-end data) computed for a set of genes or transcripts you can convert to TPM with \n",
    "\n",
    "TPM = FPKM / (sum of FPKM over all genes/transcripts) * 10^6\n",
    "\n",
    "An interesting aspect of the math here is that \n",
    "\n",
    "sum of FPKM over all transcripts = 10^6/(expression weighted mean transcript length in kb)\n",
    "\n",
    "\n",
    "And then Dewey cites his RNA-seq paper Section 1.1.1 where he gives a more detailed description of the scaling between TPM and RPKM.  Since TPM is independent of the mean transcript length, it should be more comparable between samples:\n",
    " * https://academic.oup.com/bioinformatics/article/26/4/493/243395\n",
    "\n",
    "### Experimental Differences\n",
    "\n",
    " * Gene models used\n",
    "  * Ensembl69 vs Ensembl75 vs Ensembl92\n",
    "  * Collapsed model, or individual transcripts?\n",
    " * Ribodepleted\n",
    " \n",
    " Other complications:\n",
    "  * known factors - age, gender\n",
    "  * tumour content - mixtures\n",
    "  * hidden factors - environment, temperature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Expression Effects\n",
    "\n",
    "## Genes are highly correlated in expression\n",
    "\n",
    " * Expression as a pathway property\n",
    " * Altered pathway expression - re-wiring concept\n",
    " * Measure metabolite(s) cycle activity\n",
    " * Measure Transcription Factor Levels\n",
    " * Other genomic influences:\n",
    "  * Many SNPs, copy number changes alter expression\n",
    "  * Driver fusions creating inappropriate activity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GSC RNA Expression Uses\n",
    "\n",
    "### Spearman Correlation Against Known Cancer Types\n",
    "\n",
    "Identify cancer types by gene expression patterns.\n",
    "\n",
    "![Spearman correlation plot](img/Spearman_tcga.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cibersort\n",
    "\n",
    "\n",
    "### Within Project Correlation (WPC)\n",
    "\n",
    "### Report Outliers and KB matching\n",
    "\n",
    " * Identify unusual gene expression\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-20 10:12:59 - root - INFO - logging\n",
      "2019-08-20 10:12:59 - root - INFO - logging\n",
      "2019-08-20 10:12:59 - root - INFO - Found: /home/dbleile/science/Xpress/gtex/data_cache/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct\n",
      "2019-08-20 10:12:59 - root - INFO - Found: /home/dbleile/science/Xpress/gtex/data_cache/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct\n",
      "2019-08-20 10:12:59 - root - INFO - Found: /home/dbleile/science/Xpress/gtex/data_cache/GTEx_v7_Annotations_SampleAttributesDS.txt\n",
      "2019-08-20 10:12:59 - root - INFO - Found: /home/dbleile/science/Xpress/gtex/data_cache/GTEx_Analysis_v7_Annotations_SampleAttributesDD.xlsx\n",
      "2019-08-20 10:12:59 - root - INFO - Found: /home/dbleile/science/Xpress/gtex/data_cache/GTEx_v7_Annotations_SubjectPhenotypesDS.txt\n"
     ]
    }
   ],
   "source": [
    "from gtex import GTEX_preprocess\n",
    "importlib.reload(GTEX_preprocess)\n",
    "gtex = GTEX_preprocess.GtexData()\n",
    "# gbak = GTEX_preprocess.GtexData(\"/home/dbleile/science/Xpress/gtex/data_cache_bak\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-20 10:13:07 - root - INFO - Loading reference GTEX7 row table\n",
      "2019-08-20 10:13:20 - root - INFO - loading 3 GTEX TPM rows\n",
      "2019-08-20 10:13:31 - root - INFO - Gtex TPM partial table took 0.3966263214747111 minutes\n"
     ]
    }
   ],
   "source": [
    "tp53 = gtex.gtex_gene_table(['TP53', 'PTEN'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <td>ENSG00000171862.5</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Description</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>45.42</td>\n",
       "      <td>24.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111CU-1826-SM-5GZYN</th>\n",
       "      <td>42.89</td>\n",
       "      <td>35.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111FC-0226-SM-5N9B8</th>\n",
       "      <td>40.62</td>\n",
       "      <td>40.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111VG-2326-SM-5N9BK</th>\n",
       "      <td>21.12</td>\n",
       "      <td>29.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111YS-2426-SM-5GZZQ</th>\n",
       "      <td>59.11</td>\n",
       "      <td>20.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1122O-2026-SM-5NQ91</th>\n",
       "      <td>56.43</td>\n",
       "      <td>19.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1128S-2126-SM-5H12U</th>\n",
       "      <td>35.77</td>\n",
       "      <td>25.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-113IC-0226-SM-5HL5C</th>\n",
       "      <td>43.54</td>\n",
       "      <td>35.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-117YX-2226-SM-5EGJJ</th>\n",
       "      <td>72.99</td>\n",
       "      <td>18.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11DXW-0326-SM-5H11W</th>\n",
       "      <td>27.41</td>\n",
       "      <td>32.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11DXX-2326-SM-5Q5A2</th>\n",
       "      <td>115.5</td>\n",
       "      <td>26.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11DZ1-0226-SM-5A5KF</th>\n",
       "      <td>35.85</td>\n",
       "      <td>49.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11EI6-0226-SM-5EQ64</th>\n",
       "      <td>34.32</td>\n",
       "      <td>32.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11EM3-2326-SM-5H12B</th>\n",
       "      <td>65.2</td>\n",
       "      <td>20.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11EMC-2826-SM-5PNY6</th>\n",
       "      <td>20.34</td>\n",
       "      <td>32.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11EQ8-0226-SM-5EQ5G</th>\n",
       "      <td>37.67</td>\n",
       "      <td>43.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11EQ9-2526-SM-5HL66</th>\n",
       "      <td>40.65</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11GS4-2626-SM-5A5LD</th>\n",
       "      <td>33.38</td>\n",
       "      <td>27.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11GSO-2326-SM-5A5LX</th>\n",
       "      <td>31.99</td>\n",
       "      <td>22.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11H98-0226-SM-5NQ89</th>\n",
       "      <td>44.87</td>\n",
       "      <td>36.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11I78-2626-SM-5Q5AI</th>\n",
       "      <td>61.19</td>\n",
       "      <td>30.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11LCK-1126-SM-5A5KC</th>\n",
       "      <td>57.16</td>\n",
       "      <td>37.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11O72-0226-SM-59869</th>\n",
       "      <td>31.47</td>\n",
       "      <td>39.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11OF3-2426-SM-5Q5AS</th>\n",
       "      <td>35.37</td>\n",
       "      <td>35.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11ONC-2526-SM-5986W</th>\n",
       "      <td>37.23</td>\n",
       "      <td>29.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11P7K-2026-SM-5GU74</th>\n",
       "      <td>44.62</td>\n",
       "      <td>21.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11P81-2426-SM-5GU65</th>\n",
       "      <td>42.59</td>\n",
       "      <td>29.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-11P82-1726-SM-5Q5AT</th>\n",
       "      <td>67.21</td>\n",
       "      <td>27.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZF29-0006-SM-4WKGQ</th>\n",
       "      <td>119.8</td>\n",
       "      <td>6.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZF2S-0006-SM-4WKHF</th>\n",
       "      <td>45.74</td>\n",
       "      <td>11.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZF3C-0005-SM-4WWAR</th>\n",
       "      <td>43.67</td>\n",
       "      <td>19.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZG7Y-0006-SM-4WWEA</th>\n",
       "      <td>46</td>\n",
       "      <td>4.922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZGAY-0006-SM-4WWAQ</th>\n",
       "      <td>108.5</td>\n",
       "      <td>15.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZLV1-0005-SM-4WWAP</th>\n",
       "      <td>41.46</td>\n",
       "      <td>7.778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZP4G-0006-SM-4WWE6</th>\n",
       "      <td>90.32</td>\n",
       "      <td>15.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZPCL-0006-SM-4WWAS</th>\n",
       "      <td>95.33</td>\n",
       "      <td>9.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZPIC-0005-SM-4WWEB</th>\n",
       "      <td>52.98</td>\n",
       "      <td>6.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZPU1-0006-SM-4WWAT</th>\n",
       "      <td>16.64</td>\n",
       "      <td>16.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZQG8-0005-SM-4YCEH</th>\n",
       "      <td>28.37</td>\n",
       "      <td>9.281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZQUD-0005-SM-4YCE5</th>\n",
       "      <td>42.82</td>\n",
       "      <td>5.882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZT9W-0005-SM-4YCEG</th>\n",
       "      <td>112.9</td>\n",
       "      <td>6.424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZTPG-0006-SM-4YCFG</th>\n",
       "      <td>43.52</td>\n",
       "      <td>13.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZTSS-0005-SM-4YCDR</th>\n",
       "      <td>71.71</td>\n",
       "      <td>10.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZTTD-0006-SM-51MSJ</th>\n",
       "      <td>12.99</td>\n",
       "      <td>10.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZTX8-0006-SM-4YCE4</th>\n",
       "      <td>91.34</td>\n",
       "      <td>7.632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZUA1-0005-SM-4YCEV</th>\n",
       "      <td>22.19</td>\n",
       "      <td>16.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZV68-0006-SM-4YCEJ</th>\n",
       "      <td>56.35</td>\n",
       "      <td>15.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZV7C-0005-SM-57WDL</th>\n",
       "      <td>152.3</td>\n",
       "      <td>12.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVE2-0006-SM-51MRW</th>\n",
       "      <td>42.49</td>\n",
       "      <td>3.452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVP2-0005-SM-51MRK</th>\n",
       "      <td>84.4</td>\n",
       "      <td>18.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVT2-0005-SM-57WBW</th>\n",
       "      <td>68.94</td>\n",
       "      <td>8.822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVT3-0006-SM-51MT9</th>\n",
       "      <td>18.78</td>\n",
       "      <td>10.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVT4-0006-SM-57WB8</th>\n",
       "      <td>85.23</td>\n",
       "      <td>7.498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVTK-0006-SM-57WBK</th>\n",
       "      <td>33.55</td>\n",
       "      <td>16.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVZP-0006-SM-51MSW</th>\n",
       "      <td>69.86</td>\n",
       "      <td>8.764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZVZQ-0006-SM-51MR8</th>\n",
       "      <td>18.18</td>\n",
       "      <td>13.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZXES-0005-SM-57WCB</th>\n",
       "      <td>68.68</td>\n",
       "      <td>13.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-ZXG5-0005-SM-57WCN</th>\n",
       "      <td>141.4</td>\n",
       "      <td>19.66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11690 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          0                   1\n",
       "Name                      ENSG00000171862.5  ENSG00000141510.11\n",
       "Description                            PTEN                TP53\n",
       "GTEX-1117F-0226-SM-5GZZ7              45.42               24.98\n",
       "GTEX-111CU-1826-SM-5GZYN              42.89               35.05\n",
       "GTEX-111FC-0226-SM-5N9B8              40.62               40.62\n",
       "GTEX-111VG-2326-SM-5N9BK              21.12               29.15\n",
       "GTEX-111YS-2426-SM-5GZZQ              59.11               20.39\n",
       "GTEX-1122O-2026-SM-5NQ91              56.43                19.8\n",
       "GTEX-1128S-2126-SM-5H12U              35.77               25.29\n",
       "GTEX-113IC-0226-SM-5HL5C              43.54               35.17\n",
       "GTEX-117YX-2226-SM-5EGJJ              72.99               18.94\n",
       "GTEX-11DXW-0326-SM-5H11W              27.41               32.63\n",
       "GTEX-11DXX-2326-SM-5Q5A2              115.5               26.12\n",
       "GTEX-11DZ1-0226-SM-5A5KF              35.85               49.64\n",
       "GTEX-11EI6-0226-SM-5EQ64              34.32               32.93\n",
       "GTEX-11EM3-2326-SM-5H12B               65.2                20.3\n",
       "GTEX-11EMC-2826-SM-5PNY6              20.34                32.6\n",
       "GTEX-11EQ8-0226-SM-5EQ5G              37.67               43.81\n",
       "GTEX-11EQ9-2526-SM-5HL66              40.65                  17\n",
       "GTEX-11GS4-2626-SM-5A5LD              33.38               27.56\n",
       "GTEX-11GSO-2326-SM-5A5LX              31.99               22.77\n",
       "GTEX-11H98-0226-SM-5NQ89              44.87               36.67\n",
       "GTEX-11I78-2626-SM-5Q5AI              61.19                30.1\n",
       "GTEX-11LCK-1126-SM-5A5KC              57.16               37.39\n",
       "GTEX-11O72-0226-SM-59869              31.47               39.21\n",
       "GTEX-11OF3-2426-SM-5Q5AS              35.37               35.51\n",
       "GTEX-11ONC-2526-SM-5986W              37.23               29.21\n",
       "GTEX-11P7K-2026-SM-5GU74              44.62               21.13\n",
       "GTEX-11P81-2426-SM-5GU65              42.59               29.03\n",
       "GTEX-11P82-1726-SM-5Q5AT              67.21               27.05\n",
       "...                                     ...                 ...\n",
       "GTEX-ZF29-0006-SM-4WKGQ               119.8                6.28\n",
       "GTEX-ZF2S-0006-SM-4WKHF               45.74               11.02\n",
       "GTEX-ZF3C-0005-SM-4WWAR               43.67               19.77\n",
       "GTEX-ZG7Y-0006-SM-4WWEA                  46               4.922\n",
       "GTEX-ZGAY-0006-SM-4WWAQ               108.5               15.07\n",
       "GTEX-ZLV1-0005-SM-4WWAP               41.46               7.778\n",
       "GTEX-ZP4G-0006-SM-4WWE6               90.32               15.48\n",
       "GTEX-ZPCL-0006-SM-4WWAS               95.33                9.89\n",
       "GTEX-ZPIC-0005-SM-4WWEB               52.98                6.77\n",
       "GTEX-ZPU1-0006-SM-4WWAT               16.64                16.8\n",
       "GTEX-ZQG8-0005-SM-4YCEH               28.37               9.281\n",
       "GTEX-ZQUD-0005-SM-4YCE5               42.82               5.882\n",
       "GTEX-ZT9W-0005-SM-4YCEG               112.9               6.424\n",
       "GTEX-ZTPG-0006-SM-4YCFG               43.52               13.43\n",
       "GTEX-ZTSS-0005-SM-4YCDR               71.71               10.77\n",
       "GTEX-ZTTD-0006-SM-51MSJ               12.99               10.55\n",
       "GTEX-ZTX8-0006-SM-4YCE4               91.34               7.632\n",
       "GTEX-ZUA1-0005-SM-4YCEV               22.19               16.41\n",
       "GTEX-ZV68-0006-SM-4YCEJ               56.35               15.13\n",
       "GTEX-ZV7C-0005-SM-57WDL               152.3               12.85\n",
       "GTEX-ZVE2-0006-SM-51MRW               42.49               3.452\n",
       "GTEX-ZVP2-0005-SM-51MRK                84.4               18.51\n",
       "GTEX-ZVT2-0005-SM-57WBW               68.94               8.822\n",
       "GTEX-ZVT3-0006-SM-51MT9               18.78                10.2\n",
       "GTEX-ZVT4-0006-SM-57WB8               85.23               7.498\n",
       "GTEX-ZVTK-0006-SM-57WBK               33.55                16.8\n",
       "GTEX-ZVZP-0006-SM-51MSW               69.86               8.764\n",
       "GTEX-ZVZQ-0006-SM-51MR8               18.18               13.85\n",
       "GTEX-ZXES-0005-SM-57WCB               68.68               13.14\n",
       "GTEX-ZXG5-0005-SM-57WCN               141.4               19.66\n",
       "\n",
       "[11690 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tp53.head().T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-20 10:13:47 - root - INFO - 31 separate tissue types - total samples 15598\n",
      "2019-08-20 10:13:47 - root - INFO - Adipose Tissue (n=886)\n",
      "2019-08-20 10:13:47 - root - INFO - \tAdipose - Subcutaneous (n=523)\n",
      "2019-08-20 10:13:47 - root - INFO - \tAdipose - Visceral (Omentum) (n=363)\n",
      "2019-08-20 10:13:47 - root - INFO - Adrenal Gland (n=205)\n",
      "2019-08-20 10:13:47 - root - INFO - Bladder (n=11)\n",
      "2019-08-20 10:13:47 - root - INFO - Blood (n=2561)\n",
      "2019-08-20 10:13:47 - root - INFO - \tCells - EBV-transformed lymphocytes (n=149)\n",
      "2019-08-20 10:13:47 - root - INFO - \tWhole Blood (n=2412)\n",
      "2019-08-20 10:13:47 - root - INFO - Blood Vessel (n=1041)\n",
      "2019-08-20 10:13:47 - root - INFO - \tArtery - Aorta (n=315)\n",
      "2019-08-20 10:13:47 - root - INFO - \tArtery - Coronary (n=183)\n",
      "2019-08-20 10:13:47 - root - INFO - \tArtery - Tibial (n=543)\n",
      "2019-08-20 10:13:47 - root - INFO - Bone Marrow (n=160)\n",
      "2019-08-20 10:13:47 - root - INFO - Brain (n=2076)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Amygdala (n=124)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Anterior cingulate cortex (BA24) (n=148)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Caudate (basal ganglia) (n=199)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Cerebellar Hemisphere (n=172)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Cerebellum (n=215)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Cortex (n=195)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Frontal Cortex (BA9) (n=164)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Hippocampus (n=153)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Hypothalamus (n=155)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Nucleus accumbens (basal ganglia) (n=180)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Putamen (basal ganglia) (n=149)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Spinal cord (cervical c-1) (n=113)\n",
      "2019-08-20 10:13:47 - root - INFO - \tBrain - Substantia nigra (n=109)\n",
      "2019-08-20 10:13:47 - root - INFO - Breast (n=306)\n",
      "2019-08-20 10:13:47 - root - INFO - Cervix Uteri (n=11)\n",
      "2019-08-20 10:13:47 - root - INFO - \tCervix - Ectocervix (n=6)\n",
      "2019-08-20 10:13:47 - root - INFO - \tCervix - Endocervix (n=5)\n",
      "2019-08-20 10:13:47 - root - INFO - Colon (n=539)\n",
      "2019-08-20 10:13:47 - root - INFO - \tColon - Sigmoid (n=254)\n",
      "2019-08-20 10:13:47 - root - INFO - \tColon - Transverse (n=285)\n",
      "2019-08-20 10:13:47 - root - INFO - Esophagus (n=1111)\n",
      "2019-08-20 10:13:47 - root - INFO - \tEsophagus - Gastroesophageal Junction (n=263)\n",
      "2019-08-20 10:13:47 - root - INFO - \tEsophagus - Mucosa (n=444)\n",
      "2019-08-20 10:13:47 - root - INFO - \tEsophagus - Muscularis (n=404)\n",
      "2019-08-20 10:13:47 - root - INFO - Fallopian Tube (n=7)\n",
      "2019-08-20 10:13:47 - root - INFO - Heart (n=727)\n",
      "2019-08-20 10:13:47 - root - INFO - \tHeart - Atrial Appendage (n=310)\n",
      "2019-08-20 10:13:47 - root - INFO - \tHeart - Left Ventricle (n=417)\n",
      "2019-08-20 10:13:47 - root - INFO - Kidney (n=50)\n",
      "2019-08-20 10:13:47 - root - INFO - Liver (n=188)\n",
      "2019-08-20 10:13:47 - root - INFO - Lung (n=607)\n",
      "2019-08-20 10:13:47 - root - INFO - Muscle (n=718)\n",
      "2019-08-20 10:13:47 - root - INFO - Nerve (n=498)\n",
      "2019-08-20 10:13:47 - root - INFO - Ovary (n=138)\n",
      "2019-08-20 10:13:47 - root - INFO - Pancreas (n=268)\n",
      "2019-08-20 10:13:47 - root - INFO - Pituitary (n=192)\n",
      "2019-08-20 10:13:47 - root - INFO - Prostate (n=159)\n",
      "2019-08-20 10:13:47 - root - INFO - Salivary Gland (n=104)\n",
      "2019-08-20 10:13:47 - root - INFO - Skin (n=1362)\n",
      "2019-08-20 10:13:47 - root - INFO - \tCells - Transformed fibroblasts (n=362)\n",
      "2019-08-20 10:13:47 - root - INFO - \tSkin - Not Sun Exposed (Suprapubic) (n=408)\n",
      "2019-08-20 10:13:47 - root - INFO - \tSkin - Sun Exposed (Lower leg) (n=592)\n",
      "2019-08-20 10:13:47 - root - INFO - Small Intestine (n=143)\n",
      "2019-08-20 10:13:47 - root - INFO - Spleen (n=169)\n",
      "2019-08-20 10:13:47 - root - INFO - Stomach (n=272)\n",
      "2019-08-20 10:13:47 - root - INFO - Testis (n=284)\n",
      "2019-08-20 10:13:47 - root - INFO - Thyroid (n=564)\n",
      "2019-08-20 10:13:47 - root - INFO - Uterus (n=117)\n",
      "2019-08-20 10:13:47 - root - INFO - Vagina (n=124)\n"
     ]
    }
   ],
   "source": [
    "td = gtex.tissues()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-20 10:15:56 - root - INFO - loading all rows\n",
      "2019-08-20 10:19:41 - root - INFO - Gtex TPM partial table took 3.757295882701874 minutes\n"
     ]
    }
   ],
   "source": [
    "gtpm = gtex.gtex_gene_table(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <th>GTEX-111CU-1826-SM-5GZYN</th>\n",
       "      <th>GTEX-111FC-0226-SM-5N9B8</th>\n",
       "      <th>GTEX-111VG-2326-SM-5N9BK</th>\n",
       "      <th>GTEX-111YS-2426-SM-5GZZQ</th>\n",
       "      <th>GTEX-1122O-2026-SM-5NQ91</th>\n",
       "      <th>GTEX-1128S-2126-SM-5H12U</th>\n",
       "      <th>GTEX-113IC-0226-SM-5HL5C</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZVE2-0006-SM-51MRW</th>\n",
       "      <th>GTEX-ZVP2-0005-SM-51MRK</th>\n",
       "      <th>GTEX-ZVT2-0005-SM-57WBW</th>\n",
       "      <th>GTEX-ZVT3-0006-SM-51MT9</th>\n",
       "      <th>GTEX-ZVT4-0006-SM-57WB8</th>\n",
       "      <th>GTEX-ZVTK-0006-SM-57WBK</th>\n",
       "      <th>GTEX-ZVZP-0006-SM-51MSW</th>\n",
       "      <th>GTEX-ZVZQ-0006-SM-51MR8</th>\n",
       "      <th>GTEX-ZXES-0005-SM-57WCB</th>\n",
       "      <th>GTEX-ZXG5-0005-SM-57WCN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.11580</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.02329</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.04641</td>\n",
       "      <td>0.03076</td>\n",
       "      <td>0.09358</td>\n",
       "      <td>...</td>\n",
       "      <td>0.09012</td>\n",
       "      <td>0.14620</td>\n",
       "      <td>0.10450</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.6603</td>\n",
       "      <td>0.69500</td>\n",
       "      <td>0.12130</td>\n",
       "      <td>0.41690</td>\n",
       "      <td>0.23550</td>\n",
       "      <td>0.14500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>11.03000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>8.17200</td>\n",
       "      <td>7.65800</td>\n",
       "      <td>9.37200</td>\n",
       "      <td>10.08000</td>\n",
       "      <td>13.56000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.92600</td>\n",
       "      <td>13.13000</td>\n",
       "      <td>5.53700</td>\n",
       "      <td>5.78900</td>\n",
       "      <td>8.4390</td>\n",
       "      <td>7.84300</td>\n",
       "      <td>12.39000</td>\n",
       "      <td>12.53000</td>\n",
       "      <td>8.02700</td>\n",
       "      <td>12.76000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485.2</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.06433</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.05864</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.13670</td>\n",
       "      <td>0.20790</td>\n",
       "      <td>...</td>\n",
       "      <td>0.08008</td>\n",
       "      <td>0.03607</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.10590</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.06432</td>\n",
       "      <td>0.05388</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.04756</td>\n",
       "      <td>0.05367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.03260</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.08610</td>\n",
       "      <td>0.13100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.06818</td>\n",
       "      <td>0.07309</td>\n",
       "      <td>0.03336</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.08105</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.05304</td>\n",
       "      <td>0.02996</td>\n",
       "      <td>0.03381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020.2</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.11080</td>\n",
       "      <td>0.05619</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.08739</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.04353</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 11690 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Name Description  GTEX-1117F-0226-SM-5GZZ7  \\\n",
       "0  ENSG00000223972.4     DDX11L1                   0.10820   \n",
       "1  ENSG00000227232.4      WASH7P                  21.40000   \n",
       "2  ENSG00000243485.2  MIR1302-11                   0.16020   \n",
       "3  ENSG00000237613.2     FAM138A                   0.05045   \n",
       "4  ENSG00000268020.2      OR4G4P                   0.00000   \n",
       "\n",
       "   GTEX-111CU-1826-SM-5GZYN  GTEX-111FC-0226-SM-5N9B8  \\\n",
       "0                   0.11580                   0.02104   \n",
       "1                  11.03000                  16.75000   \n",
       "2                   0.06433                   0.04674   \n",
       "3                   0.00000                   0.02945   \n",
       "4                   0.00000                   0.00000   \n",
       "\n",
       "   GTEX-111VG-2326-SM-5N9BK  GTEX-111YS-2426-SM-5GZZQ  \\\n",
       "0                   0.02329                   0.00000   \n",
       "1                   8.17200                   7.65800   \n",
       "2                   0.00000                   0.05864   \n",
       "3                   0.03260                   0.00000   \n",
       "4                   0.00000                   0.00000   \n",
       "\n",
       "   GTEX-1122O-2026-SM-5NQ91  GTEX-1128S-2126-SM-5H12U  \\\n",
       "0                   0.04641                   0.03076   \n",
       "1                   9.37200                  10.08000   \n",
       "2                   0.00000                   0.13670   \n",
       "3                   0.00000                   0.08610   \n",
       "4                   0.00000                   0.11080   \n",
       "\n",
       "   GTEX-113IC-0226-SM-5HL5C  ...  GTEX-ZVE2-0006-SM-51MRW  \\\n",
       "0                   0.09358  ...                  0.09012   \n",
       "1                  13.56000  ...                  3.92600   \n",
       "2                   0.20790  ...                  0.08008   \n",
       "3                   0.13100  ...                  0.00000   \n",
       "4                   0.05619  ...                  0.00000   \n",
       "\n",
       "   GTEX-ZVP2-0005-SM-51MRK  GTEX-ZVT2-0005-SM-57WBW  GTEX-ZVT3-0006-SM-51MT9  \\\n",
       "0                  0.14620                  0.10450                  0.00000   \n",
       "1                 13.13000                  5.53700                  5.78900   \n",
       "2                  0.03607                  0.00000                  0.10590   \n",
       "3                  0.06818                  0.07309                  0.03336   \n",
       "4                  0.00000                  0.00000                  0.00000   \n",
       "\n",
       "   GTEX-ZVT4-0006-SM-57WB8  GTEX-ZVTK-0006-SM-57WBK  GTEX-ZVZP-0006-SM-51MSW  \\\n",
       "0                   0.6603                  0.69500                  0.12130   \n",
       "1                   8.4390                  7.84300                 12.39000   \n",
       "2                   0.0000                  0.06432                  0.05388   \n",
       "3                   0.0000                  0.08105                  0.00000   \n",
       "4                   0.0000                  0.00000                  0.08739   \n",
       "\n",
       "   GTEX-ZVZQ-0006-SM-51MR8  GTEX-ZXES-0005-SM-57WCB  GTEX-ZXG5-0005-SM-57WCN  \n",
       "0                  0.41690                  0.23550                  0.14500  \n",
       "1                 12.53000                  8.02700                 12.76000  \n",
       "2                  0.00000                  0.04756                  0.05367  \n",
       "3                  0.05304                  0.02996                  0.03381  \n",
       "4                  0.00000                  0.00000                  0.04353  \n",
       "\n",
       "[5 rows x 11690 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtpm = gtmp\n",
    "gtpm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
