{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genomic Report - Main Sections\n",
    "\n",
    "## Config Generation\n",
    "## Report Generation\n",
    "\n",
    "### Data Finding\n",
    "\n",
    "### Image Finding and Transfering\n",
    "\n",
    "### Data Cleaning & Normalize\n",
    "\n",
    "### Data Combine and Organize by Gene\n",
    "\n",
    "### Match Events to KB\n",
    "\n",
    "### Group and Filter Results\n",
    "\n",
    "## Write out tables\n",
    "\n",
    "## Tracking Results\n",
    "\n",
    "## IPR upload - (Tables and Figures)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config Generation\n",
    "\n",
    " * Currently GSC specific file lookups\n",
    " * Warnings for Sarah and Wei (Image choice)\n",
    " * User editable\n",
    "\n",
    "Currently:\n",
    " * paths to images\n",
    " * paths to some data files and tables\n",
    "\n",
    "#### Considerations:  Reproducibility\n",
    " * Can we re-run with the same config and produce the same results?\n",
    "  * GraphKb version be locked?\n",
    "  * bioapps parameters - comparators - stored and frozen\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Finding\n",
    "\n",
    "Large part of the begining of create_report is being used for backwards compatibility.\n",
    "\n",
    "### Image Finding and Transfering\n",
    "\n",
    "CentralFigs - Largely untouched for JReport\n",
    " * copy and resize - image proc. dependency\n",
    " \n",
    " ### Data Cleaning & Normalize\n",
    " \n",
    "  * Gene name matching and cleaning\n",
    "  * Many imput formats for same / similar data\n",
    "  * Categorize Expression levels - categories are used for matching\n",
    "   * Use Disease and tissue comparators\n",
    "   * Mostly high / low\n",
    "  * Categorize copy number - categories are used for matching\n",
    "   * abs copy number vs copy number change / ploidy.\n",
    "   * Features like LOH local details - not used in filtering\n",
    "\n",
    "Housekeeping like - libraries -> biopsies -> sources.\n",
    "\n",
    "\n",
    "More generally:\n",
    " * Convert essential data to standard, comparable formats and categories\n",
    " * Keep other details for user analysis and detailed tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Combine and Organize by Gene\n",
    "\n",
    "Major theme - Users want one dominate repr of each type, with the ability to see other details for complicated filering.\n",
    "\n",
    "Current Measurement Groups\n",
    "\n",
    "#### Small Mutations\n",
    " * Point mutations\n",
    " * indels\n",
    " \n",
    " Input formats:\n",
    "  * hgvs directly\n",
    "  * dataSNV - INDEL merge - tab delimited\n",
    "  * MAVIS < 15bp\n",
    "  * Probes\n",
    "  * others?  cDNA? genomic? VCF?\n",
    "\n",
    "Often the same, or very similar data from all sources/tools\n",
    "\n",
    "#### Copy Number Variants\n",
    " * simple categories used with many local details.  (LOH, CNV_region)\n",
    " * Often of interest with expression.\n",
    " \n",
    "#### Expression Variants\n",
    " * only simple final categories used.\n",
    " * Many details and images\n",
    " * Many types, comparators, (mRNA vs Proteomics)\n",
    "Similar problem to Small Mutations - outlier by which comparator?  Multiple, legitimate conflicts possible\n",
    "\n",
    "#### Structural Variants\n",
    " * Probing\n",
    " * GV\n",
    " * MAVIS (larger)\n",
    "\n",
    "Difficult to combine details like DNA / RNA matching & support.\n",
    "Other details like related images."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Match Events to KB\n",
    "\n",
    " * Adds details KB details to the event.\n",
    " * Complications are factors like loose match / LOF matches\n",
    "  * Related but unsure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Group and Filter Results\n",
    "\n",
    "Categories are decided by analysts:\n",
    " * Approved therapies for this cancer\n",
    " * Approved in other cancer\n",
    " * Other therapies\n",
    " * Prognostic\n",
    " * Diagonist\n",
    " * Uncharacterized - (Related interest)\n",
    " * Novel - (For manual review)\n",
    " * (Excluded - post mortem - why missed)\n",
    "\n",
    "Categories are Ranked and mutually exclusive.\n",
    "\n",
    "Knowledgebase matches table shows these results showing (mostly) standard data.\n",
    "Oraganized by:\n",
    " * Header of event (eg. SV std rep.) with gene-level details (copy number - std rep) and sample source.\n",
    " * Matched KB event details in dropdown.\n",
    "\n",
    "Detailed tables - Subcategorize by type of event (Small mut, SV, expression, copy number)\n",
    " * These are actually GSC (experimental parameter) specific details about events above.\n",
    " * Includes a few custom analysis & combinations:\n",
    "  * SV expressed w/ DNA support\n",
    "  * combination - lowly expressed TSG with copy loss.\n",
    "\n",
    "## Write out tables\n",
    " * Perminate record step\n",
    " * Users like to read data.\n",
    " * Currently requried - complete stamp point - ready for tracking / IPR\n",
    "\n",
    "Major Refactor of tables.py:\n",
    " * Currently many details derived as needed, (eg. gene 'location'), could have been precalculated.\n",
    " \n",
    "Summarized factors are calculated after\n",
    " * counts of 'significant' events - front-page details.\n",
    " * which genes for expression images.\n",
    "\n",
    "## Tracking Results\n",
    "\n",
    "## IPR upload - (Tables and Figures)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Class\n",
    "\n",
    "Attempt to organized Data - contains all data for filter / veiw later\n",
    "Standard Gene properties:\n",
    " * symbol\n",
    " * ensembl\n",
    " * hgnc id\n",
    " * biotype - protein coding\n",
    " * other - genomic location?  Other repr - Refseq?\n",
    "\n",
    "GKB data - (without matching org.)\n",
    " * All GraphKb variants of Gene / feature\n",
    " * All GraphKb statements\n",
    "  * Can immediately see if amplifications, trials etc are relevent to gene.\n",
    "\n",
    "Measurements - Small mutations, copy number, expression, struct. var.\n",
    " * Lists of these measurement classes.\n",
    " * Can 'view' - which are 'GKB match' or 'GKB loose match' available.\n",
    "\n",
    "# Measurement Class\n",
    "\n",
    "## Small Mutation\n",
    " * Data in std form for matching - from loaders.\n",
    "  * std. hgvs notation of event\n",
    "  * categories (LOF, GOF, mutation) for GKB matching.\n",
    " * Original custom 'details' dict kept for output.\n",
    "  * Sample source.  Org. filename.\n",
    "  * Probing even? Mavis event? Delly?\n",
    " * Knows to set GKB matching if given GKB object.\n",
    "  * Statements by match level - 'match', 'postitional overlap', etc.\n",
    " * some reasonable str reps\n",
    " \n",
    "# Genome Class\n",
    " * wrapper for controlling sets of genes.\n",
    " * filters / view for table groupings.\n",
    "  * Find Genes w measurements of Approved therapies only."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
