{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GTEx v7 - Download and Basic library preparation\n",
    "\n",
    "Downloads the files.  Shows data and some plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Graphing library imports\n",
    "\n",
    "import os, sys, glob, logging\n",
    "import inspect, urllib\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from pprint import pprint\n",
    "\n",
    "import time, datetime\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# __file__ does not work in a jupyter notebook\n",
    "# make something equivalent - actually only for this particular cell\n",
    "import inspect\n",
    "# __file__ does not work in a jupyter notebook\n",
    "# make something equivalent - actually only for this particular cell\n",
    "__file__ = os.path.abspath(inspect.getfile(lambda: None))\n",
    "import pathlib\n",
    "script_dir = pathlib.Path().resolve()\n",
    "today = datetime.date.today().strftime('%Y%m%d')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-21 23:14:06 - root - INFO - logging\n",
      "/home/zeno/miniconda3/envs/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (32,34,38,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "2019-08-21 23:14:08 - root - INFO - 31 separate tissue types - total samples 15598\n",
      "2019-08-21 23:14:08 - root - INFO - Adipose Tissue (n=886)\n",
      "2019-08-21 23:14:08 - root - INFO - \tAdipose - Subcutaneous (n=523)\n",
      "2019-08-21 23:14:08 - root - INFO - \tAdipose - Visceral (Omentum) (n=363)\n",
      "2019-08-21 23:14:08 - root - INFO - Adrenal Gland (n=205)\n",
      "2019-08-21 23:14:08 - root - INFO - Bladder (n=11)\n",
      "2019-08-21 23:14:08 - root - INFO - Blood (n=2561)\n",
      "2019-08-21 23:14:08 - root - INFO - \tCells - EBV-transformed lymphocytes (n=149)\n",
      "2019-08-21 23:14:09 - root - INFO - \tWhole Blood (n=2412)\n",
      "2019-08-21 23:14:09 - root - INFO - Blood Vessel (n=1041)\n",
      "2019-08-21 23:14:09 - root - INFO - \tArtery - Aorta (n=315)\n",
      "2019-08-21 23:14:09 - root - INFO - \tArtery - Coronary (n=183)\n",
      "2019-08-21 23:14:09 - root - INFO - \tArtery - Tibial (n=543)\n",
      "2019-08-21 23:14:09 - root - INFO - Bone Marrow (n=160)\n",
      "2019-08-21 23:14:09 - root - INFO - Brain (n=2076)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Amygdala (n=124)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Anterior cingulate cortex (BA24) (n=148)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Caudate (basal ganglia) (n=199)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Cerebellar Hemisphere (n=172)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Cerebellum (n=215)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Cortex (n=195)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Frontal Cortex (BA9) (n=164)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Hippocampus (n=153)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Hypothalamus (n=155)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Nucleus accumbens (basal ganglia) (n=180)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Putamen (basal ganglia) (n=149)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Spinal cord (cervical c-1) (n=113)\n",
      "2019-08-21 23:14:09 - root - INFO - \tBrain - Substantia nigra (n=109)\n",
      "2019-08-21 23:14:09 - root - INFO - Breast (n=306)\n",
      "2019-08-21 23:14:09 - root - INFO - Cervix Uteri (n=11)\n",
      "2019-08-21 23:14:09 - root - INFO - \tCervix - Ectocervix (n=6)\n",
      "2019-08-21 23:14:09 - root - INFO - \tCervix - Endocervix (n=5)\n",
      "2019-08-21 23:14:09 - root - INFO - Colon (n=539)\n",
      "2019-08-21 23:14:09 - root - INFO - \tColon - Sigmoid (n=254)\n",
      "2019-08-21 23:14:09 - root - INFO - \tColon - Transverse (n=285)\n",
      "2019-08-21 23:14:09 - root - INFO - Esophagus (n=1111)\n",
      "2019-08-21 23:14:09 - root - INFO - \tEsophagus - Gastroesophageal Junction (n=263)\n",
      "2019-08-21 23:14:09 - root - INFO - \tEsophagus - Mucosa (n=444)\n",
      "2019-08-21 23:14:09 - root - INFO - \tEsophagus - Muscularis (n=404)\n",
      "2019-08-21 23:14:09 - root - INFO - Fallopian Tube (n=7)\n",
      "2019-08-21 23:14:09 - root - INFO - Heart (n=727)\n",
      "2019-08-21 23:14:09 - root - INFO - \tHeart - Atrial Appendage (n=310)\n",
      "2019-08-21 23:14:09 - root - INFO - \tHeart - Left Ventricle (n=417)\n",
      "2019-08-21 23:14:09 - root - INFO - Kidney (n=50)\n",
      "2019-08-21 23:14:09 - root - INFO - Liver (n=188)\n",
      "2019-08-21 23:14:09 - root - INFO - Lung (n=607)\n",
      "2019-08-21 23:14:09 - root - INFO - Muscle (n=718)\n",
      "2019-08-21 23:14:09 - root - INFO - Nerve (n=498)\n",
      "2019-08-21 23:14:09 - root - INFO - Ovary (n=138)\n",
      "2019-08-21 23:14:09 - root - INFO - Pancreas (n=268)\n",
      "2019-08-21 23:14:09 - root - INFO - Pituitary (n=192)\n",
      "2019-08-21 23:14:09 - root - INFO - Prostate (n=159)\n",
      "2019-08-21 23:14:09 - root - INFO - Salivary Gland (n=104)\n",
      "2019-08-21 23:14:09 - root - INFO - Skin (n=1362)\n",
      "2019-08-21 23:14:09 - root - INFO - \tCells - Transformed fibroblasts (n=362)\n",
      "2019-08-21 23:14:09 - root - INFO - \tSkin - Not Sun Exposed (Suprapubic) (n=408)\n",
      "2019-08-21 23:14:09 - root - INFO - \tSkin - Sun Exposed (Lower leg) (n=592)\n",
      "2019-08-21 23:14:09 - root - INFO - Small Intestine (n=143)\n",
      "2019-08-21 23:14:09 - root - INFO - Spleen (n=169)\n",
      "2019-08-21 23:14:09 - root - INFO - Stomach (n=272)\n",
      "2019-08-21 23:14:09 - root - INFO - Testis (n=284)\n",
      "2019-08-21 23:14:09 - root - INFO - Thyroid (n=564)\n",
      "2019-08-21 23:14:09 - root - INFO - Uterus (n=117)\n",
      "2019-08-21 23:14:09 - root - INFO - Vagina (n=124)\n"
     ]
    }
   ],
   "source": [
    "# Data files - Find the source GTEX data \n",
    "# Download to a cache\n",
    "SIMPLE_FORMAT = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'\n",
    "DATEFMT = '%Y-%m-%d %H:%M:%S'\n",
    "logging.basicConfig(level=logging.INFO, format=SIMPLE_FORMAT, datefmt=DATEFMT)\n",
    "logging.info(\"logging\")\n",
    "\n",
    "# GTEx data from https://gtexportal.org/home/datasets\n",
    "# Gene transcripts per million data\n",
    "cache_dir_alt = os.path.join(pathlib.Path().resolve(), 'gtex', 'data_cache')\n",
    "cache_dir = '/gsc/resources/expression/GTEx/v7'\n",
    "if not os.path.exists(cache_dir) and os.path.exists(cache_dir_alt):\n",
    "    cache_dir = cache_dir_alt\n",
    "if not os.path.exists(cache_dir):\n",
    "    logging.info('Creating {}'.format(cache_dir))\n",
    "    os.makedirs(cache_dir, mode=0o777)\n",
    "\n",
    "hgnc_fns = glob.glob(os.path.join(cache_dir, 'hgnc*'))\n",
    "if not hgnc_fns:\n",
    "    hgnc_fn = os.path.join(cache_dir, 'hgnc_{}.tsv'.format(\n",
    "        datetime.date.today().strftime('%Y%m%d')))\n",
    "    logging.info('Downloading HGNC annotations to: %s', hgnc_fn)\n",
    "    urllib.request.urlretrieve(\n",
    "        \"ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt\", hgnc_fn)\n",
    "hgnc_fns = glob.glob(os.path.join(cache_dir, 'hgnc*'))\n",
    "hgnc = pd.read_csv(hgnc_fns[0], sep='\\t')\n",
    "\n",
    "# Detailed data file:\n",
    "# This file is quite large ~ 1GB\n",
    "gtex_tpm_link = 'https://storage.googleapis.com/gtex_analysis_v7/rna_seq_data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct.gz'\n",
    "\n",
    "# Try Using unzipped version\n",
    "gtex_tpm_fn = os.path.join(cache_dir, 'GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct')\n",
    "if not os.path.exists(gtex_tpm_fn):\n",
    "    gtex_tpm_fn = os.path.join(cache_dir, 'GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct.gz')\n",
    "if not os.path.exists(gtex_tpm_fn):\n",
    "    print(\"downloading {}\".format(gtex_tpm_link))\n",
    "    urllib.request.urlretrieve(gtex_tpm_link, gtex_tpm_fn)\n",
    "\n",
    "# Just looking at median values by tissue (SMTSD) greatly reduces file size\n",
    "gtex_tpm_med_link = 'https://storage.googleapis.com/gtex_analysis_v7/rna_seq_data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct.gz'\n",
    "# Try using unzipped version\n",
    "gtex_tpm_med_fn = os.path.join(cache_dir, 'GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct')\n",
    "if not os.path.exists(gtex_tpm_med_fn):\n",
    "    gtex_tpm_med_fn = os.path.join(cache_dir, 'GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct.gz')\n",
    "if not os.path.exists(gtex_tpm_med_fn):\n",
    "    print(\"downloading {}\".format(gtex_tpm_med_link))\n",
    "    urllib.request.urlretrieve(gtex_tpm_med_link, gtex_tpm_med_fn)\n",
    "\n",
    "# Sample annotation data\n",
    "\n",
    "# Subject Phenotype data\n",
    "# Age, sex and Hardy Scale death circumstances\n",
    "gtex_pheno_link = 'https://storage.googleapis.com/gtex_analysis_v7/annotations/GTEx_v7_Annotations_SubjectPhenotypesDS.txt'\n",
    "gtex_pheno_fn = os.path.join(cache_dir, 'GTEx_v7_Annotations_SubjectPhenotypesDS.txt')\n",
    "\n",
    "# Main sample data of interest is:\n",
    "# 'SMTS': 'Tissue Type, area from which the tissue sample was taken.  This is a parent value to SMTSD.'\n",
    "# 'SMTSD': 'SMTS Detailed'\n",
    "gtex_attr_link = 'https://storage.googleapis.com/gtex_analysis_v7/annotations/GTEx_v7_Annotations_SampleAttributesDS.txt'\n",
    "gtex_attr_fn = os.path.join(cache_dir, 'GTEx_v7_Annotations_SampleAttributesDS.txt')\n",
    "gtex_attr_desc_link = 'https://storage.googleapis.com/gtex_analysis_v7/annotations/GTEx_Analysis_v7_Annotations_SampleAttributesDD.xlsx'\n",
    "gtex_attr_desc_fn = os.path.join(cache_dir, 'GTEx_Analysis_v7_Annotations_SampleAttributesDD.xlsx')\n",
    "if not os.path.exists(gtex_attr_fn):\n",
    "    logging.info('downloading {}'.format(gtex_attr_link))\n",
    "    urllib.request.urlretrieve(gtex_attr_link, gtex_attr_fn)\n",
    "if not os.path.exists(gtex_attr_desc_fn):\n",
    "    logging.info('downloading {}'.format(gtex_attr_desc_link))\n",
    "    urllib.request.urlretrieve(gtex_attr_desc_link, gtex_attr_desc_fn)\n",
    "\n",
    "# Load tissue specific detatils\n",
    "gtex_attr = pd.read_csv(gtex_attr_fn, sep='\\t')\n",
    "tissue_types = sorted(gtex_attr['SMTS'].unique())\n",
    "tissue_dict = dict()\n",
    "for tissue in tissue_types:\n",
    "    tissue_dict[tissue] = dict()\n",
    "    tissue_dict[tissue]['samples'] = gtex_attr[gtex_attr['SMTS'] == tissue]['SAMPID'].to_list()\n",
    "    tissue_dict[tissue]['subtypes'] = dict()\n",
    "    for subtype in gtex_attr[gtex_attr['SMTS'] == tissue]['SMTSD'].unique():\n",
    "        # Get all samples by subtype\n",
    "        tissue_dict[tissue]['subtypes'][subtype] = gtex_attr[gtex_attr['SMTSD'] == subtype]['SAMPID'].to_list()\n",
    "    \n",
    "\n",
    "    sorted(gtex_attr[gtex_attr['SMTS'] == tissue]['SMTSD'].unique())\n",
    "tissue_subtypes = sorted(gtex_attr['SMTSD'].unique())\n",
    "\n",
    "# log number of tissue samples\n",
    "logging.info('%s separate tissue types - total samples %s', len(tissue_dict), len(gtex_attr['SAMPID']))\n",
    "for tissue in sorted(tissue_dict.keys()):\n",
    "    logging.info('%s (n=%s)', tissue, len(tissue_dict[tissue]['samples']))\n",
    "    if len(tissue_dict[tissue]['subtypes']) > 1:\n",
    "        for subtype in sorted(tissue_dict[tissue]['subtypes'].keys()):\n",
    "            logging.info('\\t%s (n=%s)', subtype, len(tissue_dict[tissue]['subtypes'][subtype]))\n",
    "# logging.info(pprint.pformat(tissue_dict))\n",
    "\n",
    "# Other annotation data\n",
    "\n",
    "# Mutually exclusive \n",
    "# Which study was this from?\n",
    "mut_info_fn = os.path.join(cache_dir, 'enrichments-analysis-result.txt')\n",
    "#mut_info = pd.read_csv(mut_info_fn, sep='\\t')\n",
    "\n",
    "# Human housekeeping gene - stable expression\n",
    "\n",
    "## https://www.cell.com/trends/genetics/pdf/S0168-9525(13)00089-9.pdf - Human housekeeping genes, revisited\n",
    "\n",
    "# https://m.tau.ac.il/~elieis/HKG/HK_genes.txt\n",
    "# E. Eisenberg and E.Y. Levanon, Trends in Genetics, 29 (2013)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-21 23:38:39 - root - INFO - Loading Row partial table ...\n",
      "2019-08-21 23:39:06 - root - INFO - Loading column partial table ...\n"
     ]
    }
   ],
   "source": [
    "# Some table loading helper functions\n",
    "\n",
    "# The full set of all all GTEx tpms is very large and can be hard to work with\n",
    "# Some functions only load some of the data at a time.\n",
    "\n",
    "# Genes of interest - some test genes \n",
    "TEST_GENES = [\n",
    "    'TP53',\n",
    "    'ERBB2',  # Herceptin target\n",
    "    'EGFR',\n",
    "    'AKT1',\n",
    "    'KRAS',\n",
    "    'PTEN',\n",
    "    'APOE',\n",
    "]\n",
    "\n",
    "TUMOUR_MUT_GENES = [  # www.tumourportal.org - highly mutated\n",
    "    'TP53',\n",
    "    'PIK3CA',\n",
    "    'PTEN',\n",
    "    'KRAS',\n",
    "    'APC',\n",
    "    'MLL3',  # aka KMT2C\n",
    "    'KMT2C',\n",
    "    'FAT1',\n",
    "    'MLL2',  # aka KMT2D\n",
    "    'KMT2D',\n",
    "    'ARID1A',\n",
    "    'VHL',\n",
    "    'PBRM1',\n",
    "    'NF1',\n",
    "    'EGFR',\n",
    "    'ATM',\n",
    "    'PIK3R1',\n",
    "    'BRAF',\n",
    "    'CDKN2A',\n",
    "    'SETD2',\n",
    "    'CREBBP',\n",
    "    'FBXW7',\n",
    "    'SPEN',\n",
    "    'MTOR',\n",
    "    'RB1',\n",
    "    'SMARCA4',\n",
    "    'NOTCH1',\n",
    "]\n",
    "\n",
    "TEST_GENES = sorted(set(TEST_GENES + TUMOUR_MUT_GENES))\n",
    "\n",
    "ts = []\n",
    "onco = []\n",
    "with open(os.path.join(script_dir, 'tables', 'TS_genes_in_kb.txt')) as ifile:\n",
    "    for line in ifile:\n",
    "        if line.strip():\n",
    "            ts.append(line.strip())\n",
    "with open(os.path.join(script_dir, 'tables', 'oncogenes_in_kb.txt')) as ifile:\n",
    "    for line in ifile:\n",
    "        if line.strip():\n",
    "            onco.append(line.strip())\n",
    "\n",
    "ts_onco = sorted(set(ts + onco + TEST_GENES))\n",
    "\n",
    "# gtex_tpm file has genes as rows.\n",
    "# First two columns are gene information\n",
    "# Over 10000 columns of samples\n",
    "\n",
    "#                Name Description  GTEX-1117F-0226-SM-5GZZ7  GTEX-111CU-1826-SM-5GZYN ...\n",
    "#0  ENSG00000223972.4     DDX11L1                   0.10820                   0.11580\n",
    "#1  ENSG00000227232.4      WASH7P                  21.40000                  11.03000\n",
    "\n",
    "\n",
    "GTPM_SKIP = 2  #  skip two comment rows at the beginning of the file\n",
    "\n",
    "# Just getting a few rows of this giant file\n",
    "logging.info('Loading Row partial table ...')\n",
    "gtpm_row_table = pd.read_csv(gtex_tpm_fn, skiprows=2, sep='\\t', usecols=range(4))\n",
    "logging.info('Loading column partial table ...')\n",
    "gtpm_column_table = pd.read_csv(gtex_tpm_fn, skiprows=lambda x: x not in [2, 3], sep='\\t')\n",
    "\n",
    "def gtpm_row_by_genename(gene_name):\n",
    "    sub_table = gtpm_row_table[gtpm_row_table['Description'] == gene_name]\n",
    "    if sub_table.empty:\n",
    "        logging.error(\"Missing '%s row in GTEX TPM table'\", gene_name)\n",
    "        return None\n",
    "    return [GTPM_SKIP + 1 + i for i in sub_table.index][0]\n",
    "\n",
    "def gtpm_keep_rows(gene_list=TEST_GENES):\n",
    "    row_list = [gtpm_row_by_genename(gene) for gene in gene_list]\n",
    "    return [GTPM_SKIP] + [gene for gene in row_list if gene]\n",
    "\n",
    "\n",
    "def gtex_tpm_partial_table_load(gene_list, sample_list=None):\n",
    "    start_time = time.time()\n",
    "    if gene_list:\n",
    "        keep_rows = gtpm_keep_rows(gene_list)\n",
    "        skiprows_func = lambda x: x not in keep_rows\n",
    "        logging.info('loading %s GTEX TPM rows', len(keep_rows))\n",
    "    else:\n",
    "        logging.info('loading all rows')\n",
    "        skiprows_func = lambda x: x in [0, 1]\n",
    "    \n",
    "    if sample_list:\n",
    "        logging.info('Limiting load to %s potential samples', len(sample_list))\n",
    "        # Not all samples are in tables\n",
    "        sample_list = [s for s in sample_list if s in gtpm_column_table.columns]\n",
    "        logging.info('Limiting load to %s samples', len(sample_list))\n",
    "        columns = ['Name', 'Description'] + sample_list\n",
    "    else:\n",
    "        sample_list = None\n",
    "    \n",
    "    gtpm_partial = pd.read_csv(\n",
    "        gtex_tpm_fn,\n",
    "        skiprows=skiprows_func,  # Only selected gene rows\n",
    "        sep='\\t',\n",
    "        header=0,\n",
    "        usecols=sample_list,  # Number of samples limit\n",
    "    )\n",
    "    end_time = time.time()\n",
    "    logging.info(\"Gtex TPM partial table took {} minutes\".format((end_time - start_time)/60))\n",
    "    return gtpm_partial\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-21 23:16:17 - root - INFO - Loading g2 row partial table ...\n",
      "2019-08-21 23:17:09 - root - INFO - Loading g2 column partial table ...\n",
      "2019-08-21 23:18:02 - root - INFO - done\n"
     ]
    }
   ],
   "source": [
    "# Small table of median values per tissue only\n",
    "gtpm_med = pd.read_csv(gtex_tpm_med_fn, sep='\\t', skiprows=2, low_memory=False)\n",
    "# gtex log2\n",
    "g2_fn = os.path.join(cache_dir, 'preprocess', 'GTEX_log2.tsv')\n",
    "logging.info('Loading g2 row partial table ...')\n",
    "g2_row_table = pd.read_csv(g2_fn, sep='\\t', usecols=range(4))\n",
    "logging.info('Loading g2 column partial table ...')\n",
    "g2_col_table = pd.read_csv(g2_fn, skiprows=lambda x: x not in [0, 1, 2], sep='\\t')\n",
    "logging.info('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Symbol            ensid  ens_ver  GTEX-1117F-0226-SM-5GZZ7\n",
      "0     DDX11L1  ENSG00000223972        4                 -3.208228\n",
      "1      WASH7P  ENSG00000227232        4                  4.419539\n",
      "2  MIR1302-11  ENSG00000243485        2                 -2.642054\n",
      "3     FAM138A  ENSG00000237613        2                 -4.309002\n",
      "4      OR4G4P  ENSG00000268020        2                       NaN\n",
      "30214\n"
     ]
    }
   ],
   "source": [
    "g2_samples = list(g2_col_table.columns[3:])\n",
    "print(g2_row_table.head())\n",
    "#print(g2_col_table.head())\n",
    "print(len(g2_row_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2019-08-21 23:36:40 - root - INFO - Adipose Tissue (n=797)\n",
      "2019-08-21 23:36:40 - root - INFO - \tAdipose - Subcutaneous (n=442)\n",
      "2019-08-21 23:36:40 - root - INFO - \tAdipose - Visceral (Omentum) (n=355)\n",
      "2019-08-21 23:36:40 - root - INFO - Adrenal Gland (n=190)\n",
      "2019-08-21 23:36:40 - root - INFO - Bladder (n=11)\n",
      "2019-08-21 23:36:40 - root - INFO - Blood (n=537)\n",
      "2019-08-21 23:36:40 - root - INFO - \tCells - EBV-transformed lymphocytes (n=130)\n",
      "2019-08-21 23:36:40 - root - INFO - \tWhole Blood (n=407)\n",
      "2019-08-21 23:36:40 - root - INFO - Blood Vessel (n=913)\n",
      "2019-08-21 23:36:40 - root - INFO - \tArtery - Aorta (n=299)\n",
      "2019-08-21 23:36:40 - root - INFO - \tArtery - Coronary (n=173)\n",
      "2019-08-21 23:36:40 - root - INFO - \tArtery - Tibial (n=441)\n",
      "2019-08-21 23:36:40 - root - INFO - Bone Marrow (n=0)\n",
      "2019-08-21 23:36:40 - root - INFO - Brain (n=1671)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Amygdala (n=100)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Anterior cingulate cortex (BA24) (n=121)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Caudate (basal ganglia) (n=160)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Cerebellar Hemisphere (n=136)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Cerebellum (n=173)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Cortex (n=158)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Frontal Cortex (BA9) (n=129)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Hippocampus (n=123)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Hypothalamus (n=121)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Nucleus accumbens (basal ganglia) (n=147)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Putamen (basal ganglia) (n=124)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Spinal cord (cervical c-1) (n=91)\n",
      "2019-08-21 23:36:40 - root - INFO - \tBrain - Substantia nigra (n=88)\n",
      "2019-08-21 23:36:40 - root - INFO - Breast (n=290)\n",
      "2019-08-21 23:36:40 - root - INFO - Cervix Uteri (n=11)\n",
      "2019-08-21 23:36:40 - root - INFO - \tCervix - Ectocervix (n=6)\n",
      "2019-08-21 23:36:40 - root - INFO - \tCervix - Endocervix (n=5)\n",
      "2019-08-21 23:36:40 - root - INFO - Colon (n=507)\n",
      "2019-08-21 23:36:40 - root - INFO - \tColon - Sigmoid (n=233)\n",
      "2019-08-21 23:36:40 - root - INFO - \tColon - Transverse (n=274)\n",
      "2019-08-21 23:36:40 - root - INFO - Esophagus (n=1021)\n",
      "2019-08-21 23:36:40 - root - INFO - \tEsophagus - Gastroesophageal Junction (n=244)\n",
      "2019-08-21 23:36:40 - root - INFO - \tEsophagus - Mucosa (n=407)\n",
      "2019-08-21 23:36:40 - root - INFO - \tEsophagus - Muscularis (n=370)\n",
      "2019-08-21 23:36:40 - root - INFO - Fallopian Tube (n=7)\n",
      "2019-08-21 23:36:40 - root - INFO - Heart (n=600)\n",
      "2019-08-21 23:36:40 - root - INFO - \tHeart - Atrial Appendage (n=297)\n",
      "2019-08-21 23:36:40 - root - INFO - \tHeart - Left Ventricle (n=303)\n",
      "2019-08-21 23:36:40 - root - INFO - Kidney (n=45)\n",
      "2019-08-21 23:36:40 - root - INFO - Liver (n=175)\n",
      "2019-08-21 23:36:40 - root - INFO - Lung (n=427)\n",
      "2019-08-21 23:36:40 - root - INFO - Muscle (n=564)\n",
      "2019-08-21 23:36:40 - root - INFO - Nerve (n=414)\n",
      "2019-08-21 23:36:40 - root - INFO - Ovary (n=133)\n",
      "2019-08-21 23:36:40 - root - INFO - Pancreas (n=248)\n",
      "2019-08-21 23:36:40 - root - INFO - Pituitary (n=183)\n",
      "2019-08-21 23:36:40 - root - INFO - Prostate (n=152)\n",
      "2019-08-21 23:36:40 - root - INFO - Salivary Gland (n=97)\n",
      "2019-08-21 23:36:40 - root - INFO - Skin (n=1203)\n",
      "2019-08-21 23:36:40 - root - INFO - \tCells - Transformed fibroblasts (n=343)\n",
      "2019-08-21 23:36:40 - root - INFO - \tSkin - Not Sun Exposed (Suprapubic) (n=387)\n",
      "2019-08-21 23:36:40 - root - INFO - \tSkin - Sun Exposed (Lower leg) (n=473)\n",
      "2019-08-21 23:36:40 - root - INFO - Small Intestine (n=137)\n",
      "2019-08-21 23:36:40 - root - INFO - Spleen (n=162)\n",
      "2019-08-21 23:36:40 - root - INFO - Stomach (n=262)\n",
      "2019-08-21 23:36:40 - root - INFO - Testis (n=259)\n",
      "2019-08-21 23:36:40 - root - INFO - Thyroid (n=446)\n",
      "2019-08-21 23:36:40 - root - INFO - Uterus (n=111)\n",
      "2019-08-21 23:36:40 - root - INFO - Vagina (n=115)\n"
     ]
    }
   ],
   "source": [
    "g2_samples = list(g2_col_table.columns[3:])\n",
    "td = tissue_dict.copy()\n",
    "for tissue in td.keys():\n",
    "    td[tissue]['samples'] = [s for s in td[tissue]['samples'] if s in g2_samples]\n",
    "    for subtype in td[tissue]['subtypes'].keys():\n",
    "        td[tissue]['subtypes'][subtype] = [s for s in td[tissue]['subtypes'][subtype] if s in g2_samples]\n",
    "for tissue in sorted(td.keys()):\n",
    "    logging.info('%s (n=%s)', tissue, len(td[tissue]['samples']))\n",
    "    if len(td[tissue]['subtypes']) > 1:\n",
    "        for subtype in sorted(td[tissue]['subtypes'].keys()):\n",
    "            logging.info('\\t%s (n=%s)', subtype, len(td[tissue]['subtypes'][subtype]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Apply some known values\n",
    "gm['mean'] = gm.apply(lambda row: row.mean(), axis=1)\n",
    "gm['median'] = gm.apply(lambda row: row.median(), axis=1)\n",
    "gm['std'] = gm.apply(lambda row: row.std(), axis=1)\n",
    "#gm['mode'] = gm.apply(lambda row: row.mode(), axis=1)\n",
    "gm['CoV'] = gm.apply(lambda row: pandas.np.inf if row['mean'] == 0 else row['std']/row['mean'], axis=1)\n",
    "gm['sem'] = gm.apply(lambda row: row.sem(), axis=1)\n",
    "gm.head()\n",
    "\n",
    "# high expression # low deviation among tissues\n",
    "gm_bench = gm[(gm['median'] >= 1)  # high expression\n",
    "             & (gm['median'] - 3.5 * gm['std'] > 0)\n",
    "            ]\n",
    "\n",
    "#cross_tissue_genes = sorted(gm_bench['Description'].tolist())\n",
    "\n",
    "print(\"Of {} genes, the top {} in stable cross-tissue expression are:\".format(len(gm), len(gm_bench)))\n",
    "#for gene in cross_tissue_genes:\n",
    "#    print('\\t{}'.format(gene))\n",
    "gm_bench.sort_values(by='CoV', inplace=True)\n",
    "gm_bench\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gm.sort_values(by='sem').head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gm.loc[TEST_GENES]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transpose to get Gene Columns\n",
    "gt = gtpm[gtpm.columns[1:]]\n",
    "gt.set_index('Description', inplace=True)\n",
    "org_cols = gt.columns\n",
    "\n",
    "#gtt = gt.transpose()\n",
    "#gtt.head()\n",
    "gt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gt.loc['TP53'].describe())\n",
    "print(gt.loc['ERBB2'].describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gt.loc['TP53'].value_counts(bins=100, sort=False)\n",
    "#gt.loc['PTEN'].skew()\n",
    "#gt.loc['PTEN'].kurtosis()\n",
    "gt.loc['PTEN'].quantile(0.9)\n",
    "gt.T[TEST_GENES[:-1]].iplot(kind='box',)\n",
    "gt.T[['TP53', 'PTEN']].plot(kind='box',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene = 'TP53'\n",
    "gene = 'PTEN'\n",
    "title = '{}: Samples {}'.format(gene, len(gt.loc[gene]))\n",
    "#gtt[gene].iplot(kind='hist', title=title, yTitle='counts', xTitle='Expression (TPM)', bins=100)\n",
    "# gt.loc[gene].iplot(kind='hist', title=title, yTitle='counts', xTitle='Expression (TPM)', bins=100)\n",
    "# gt.loc[gene].plot(kind='hist', title=title, bins=100)\n",
    "\n",
    "sns.set()\n",
    "fig = plt.figure()\n",
    "plt.ylabel('counts')\n",
    "sns.distplot(gt.loc[gene], kde=False, rug=False)\n",
    "out_filename = os.path.join(script_dir, 'img', gene + '_raw.png')\n",
    "plt.title(gene + ' skew: {} kurtosis: {}'.format(round(gt.loc[gene].skew(), 2),\n",
    "                                              round(gt.loc[gene].kurtosis(), 2)))\n",
    "fig.savefig(out_filename)\n",
    "fig = plt.figure()\n",
    "\n",
    "ptenlog2 = gt.loc[gene].apply(numpy.log2)\n",
    "plt.title('log2({}) skew: {} kurtosis: {}'.format(gene,\n",
    "                                                  round(ptenlog2.skew(), 2),\n",
    "                                                  round(ptenlog2.kurtosis(), 2)))\n",
    "sns.distplot(ptenlog2, kde=False, rug=False)\n",
    "plt.ylabel('counts')\n",
    "plt.xlabel(gene + ' log2')\n",
    "out_filename = os.path.join(script_dir, 'img', gene + '_log2.png')\n",
    "fig.savefig(out_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'TP53'\n",
    "\n",
    "title = '{}: Samples {}'.format(gene, len(gt.loc[gene]))\n",
    "#gtt[gene].iplot(kind='hist', title=title, yTitle='counts', xTitle='Expression (TPM)', bins=100)\n",
    "# gt.loc[gene].iplot(kind='hist', title=title, yTitle='counts', xTitle='Expression (TPM)', bins=100)\n",
    "# gt.loc[gene].plot(kind='hist', title=title, bins=100)\n",
    "\n",
    "sns.set()\n",
    "fig = plt.figure()\n",
    "plt.ylabel('counts')\n",
    "sns.distplot(gt.loc[gene], kde=False, rug=False)\n",
    "out_filename = os.path.join(script_dir, 'img', gene + '_raw.png')\n",
    "plt.title(gene + ' skew: {} kurtosis: {}'.format(round(gt.loc[gene].skew(), 2),\n",
    "                                              round(gt.loc[gene].kurtosis(), 2)))\n",
    "fig.savefig(out_filename)\n",
    "fig = plt.figure()\n",
    "\n",
    "ptenlog2 = gt.loc[gene].apply(numpy.log2)\n",
    "plt.title('log2({}) skew: {} kurtosis: {}'.format(gene,\n",
    "                                                  round(ptenlog2.skew(), 2),\n",
    "                                                  round(ptenlog2.kurtosis(), 2)))\n",
    "sns.distplot(ptenlog2, kde=False, rug=False)\n",
    "plt.ylabel('counts')\n",
    "plt.xlabel(gene + ' log2')\n",
    "out_filename = os.path.join(script_dir, 'img', gene + '_log2.png')\n",
    "fig.savefig(out_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gt.head()\n",
    "# sns.distplot(gt.loc[gene], kde=False, rug=False,)\n",
    "fig = plt.figure()\n",
    "\n",
    "for tissue in ['Heart', 'Nerve', \n",
    "               'Skin',\n",
    "               'Lung',\n",
    "               'Pancreas'\n",
    "              ]:\n",
    "    samples = tissue_dict[tissue]['samples']\n",
    "    # print(tissue)\n",
    "    # print(len(samples))\n",
    "    gt_samples = [s for s in samples if s in gt.columns]\n",
    "    dt = gt.loc[gene][gt_samples].dropna().apply(numpy.log2)\n",
    "    sns.distplot(dt, kde=True, rug=False, label=tissue)\n",
    "plt.legend()\n",
    "plt.title('log2(TP53) on some tissues')\n",
    "out_filename = os.path.join(script_dir, 'img', gene + '_by_tissue.png')\n",
    "fig.savefig(out_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(plotly.__version__)\n",
    "print(cufflinks.__version__)\n",
    "#print(ff.create_distplot.__doc__)\n",
    "gene_list = ['TP53', 'PTEN', 'KRAS', 'ERBB2', 'CDKN2A']\n",
    "gene_list = ['TP53', 'PTEN', 'KRAS',]\n",
    "fig = ff.create_distplot([gt.loc[g] for g in gene_list], gene_list)\n",
    "iplot(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# by tissue types\n",
    "gene = 'TP53'\n",
    "data_lists = []\n",
    "labels = []\n",
    "for tissue in tissue_dict.keys():\n",
    "    samples = tissue_dict[tissue]['samples']\n",
    "    # print(tissue)\n",
    "    # print(len(samples))\n",
    "    gt_samples = [s for s in samples if s in gt.columns]\n",
    "    if len(gt_samples) > 20:\n",
    "        dt = gt.loc[gene][gt_samples].dropna()\n",
    "        if not dt.empty:\n",
    "            data_lists.append(dt)\n",
    "            labels.append('{}({}) ~ {}'.format(tissue, len(samples), round(dt.median(), 1)))\n",
    "    else:\n",
    "        print('Only {} samples for \"{}\"'.format(len(gt_samples), tissue))\n",
    "# add all samples\n",
    "dt = gt.loc[gene].dropna()\n",
    "data_lists.append(dt)\n",
    "labels.append('{} - all ({}) ~ {}'.format(gene, len(dt), round(dt.median(), 1)))\n",
    "\n",
    "fig_tp53 = ff.create_distplot(data_lists, labels, show_hist=False, show_rug=False, show_curve=True)  # curve_type='normal')\n",
    "fig_tp53.layout.update(title=gene)\n",
    "iplot(fig_tp53)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Expected distributions are roughly log normal - not normal distributions.\n",
    "dir(go)\n",
    "gene='PTEN'\n",
    "tpm_label = \"TPM      {} skew: {} kurtosis: {}\".format(gene,\n",
    "                                                       round(gt.loc[gene].skew(), 2),\n",
    "                                                       round(gt.loc[gene].kurtosis(), 2))\n",
    "log2_label = \"log2TPM {} skew: {} kurtosis: {}\".format(gene,\n",
    "                                                       round(gt.loc[gene].apply(numpy.log2).skew(), 2),\n",
    "                                                       round(gt.loc[gene].apply(numpy.log2).kurtosis(), 2))\n",
    "\n",
    "fig_log2vsNormal = ff.create_distplot([gt.loc[gene].apply(numpy.log2), gt.loc[gene]],\n",
    "                         [log2_label, tpm_label],\n",
    "                         bin_size=[0.1, 2],\n",
    "                         curve_type='normal',)\n",
    "iplot(fig_log2vsNormal)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data structure\n",
    "\n",
    "Data like TPM is lognormal distributed. \n",
    " * There cannot be a negative number of transcripts.\n",
    " * log transforms (like fold-change - log 2) are more normally disturbuted.\n",
    " \n",
    " More irregular curves, like TP53 show more log-normal distributions when clustered into relevent groups, like tissue types.\n",
    " \n",
    " Otherwise, something like sklearn.preprocessing.RobustScaler looks good for IQR scaling - robust to outliers.\n",
    " \n",
    " This Scaler removes the median and scales the data according to the quantile range (defaults to IQR: Interquartile Range). The IQR is the range between the 1st quartile (25th quantile) and the 3rd quartile (75th quantile).\n",
    " \n",
    " \n",
    " ## Correlation\n",
    " \n",
    " Gene expressions are correlated / anti-correlated (mutual information/effect) in interaction pathways.  'Activation' of certiain pathways is of interest for cancer and drug effects measures.\n",
    " \n",
    " \n",
    " PCA with 'whitening' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize to log2RPKM - log normal data\n",
    "# Data is now effectively in fold change units\n",
    "\n",
    "# Use a minum value to prevent div by zero.\n",
    "#safelog2 = lambda x: numpy.log2(x) if x else numpy.log2(sys.float_info.min)\n",
    "safelog2 = lambda x: numpy.log2(x) if x else numpy.nan\n",
    "gfc = gt.applymap(safelog2)\n",
    "\n",
    "print(gfc.info())\n",
    "print(gfc.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try data reduction like PCA and LDA\n",
    "\n",
    "from sklearn.preprocessing import PowerTransformer\n",
    "from sklearn.preprocessing import RobustScaler\n",
    "\n",
    "# Which way are we scaling? Rows or columns?\n",
    "rs = RobustScaler()\n",
    "scaler = rs.fit(gfc.T)\n",
    "gfc_norm = pandas.DataFrame(scaler.transform(gfc.T), index=gfc.columns, columns=gfc.index)\n",
    "print(gfc_norm.head())\n",
    "gfc_norm.describe()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# by tissue types\n",
    "gene = 'PTEN'\n",
    "data_lists = []\n",
    "labels = []\n",
    "for tissue in tissue_dict.keys():\n",
    "    samples = tissue_dict[tissue]['samples']\n",
    "    # print(tissue)\n",
    "    # print(len(samples))\n",
    "    gt_samples = [s for s in samples if s in gfc_norm.index]\n",
    "    if len(gt_samples) > 20:\n",
    "        dt = gfc_norm[gene][gt_samples].dropna()\n",
    "        if not dt.empty:\n",
    "            data_lists.append(dt)\n",
    "            labels.append('{}({}) ~ {}'.format(tissue, len(samples), round(dt.median(), 1)))\n",
    "    else:\n",
    "        print('Only {} samples for \"{}\"'.format(len(gt_samples), tissue))\n",
    "# add all samples\n",
    "dt = gfc_norm[gene].dropna()\n",
    "data_lists.append(dt)\n",
    "labels.append('{} - all ({}) ~ {}'.format(gene, len(dt), round(dt.median(), 1)))\n",
    "\n",
    "fig_log2tissue = ff.create_distplot(data_lists, labels, show_hist=False, show_rug=False, show_curve=True)  # curve_type='normal')\n",
    "fig_log2tissue.layout.update(title=gene)\n",
    "iplot(fig_log2tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imitate https://stackabuse.com/implementing-lda-in-python-with-scikit-learn/\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import confusion_matrix, accuracy_score\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "import sklearn.decomposition\n",
    "print('sklearn.decomposition')\n",
    "pprint([f for f in dir(sklearn.decomposition) if str(f)[0] != '_'])\n",
    "\n",
    "import sklearn.discriminant_analysis\n",
    "print('sklearn.discriminant_analysis')\n",
    "pprint([f for f in dir(sklearn.discriminant_analysis) if str(f)[0] != '_'])\n",
    "\n",
    "print('sklearn.ensemble')\n",
    "import sklearn.ensemble\n",
    "pprint([f for f in dir(sklearn.ensemble) if str(f)[0] != '_'])\n",
    "\n",
    "print('sklearn.metrics')\n",
    "import sklearn.metrics\n",
    "pprint([f for f in dir(sklearn.metrics) if str(f)[0] != '_'])\n",
    "pprint(dir(sklearn.metrics))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCA\n",
    "# sparse PCA is probably better - interpret which genes are more important.\n",
    "\n",
    "pca3 = PCA(n_components=3)  # Alt opt PCA(0.99) - enough components for 99% of variance\n",
    "sparse_pca = sklearn.decomposition.SparsePCA(n_components=3, normalize_components=True)\n",
    "sparse_comp = sparse_pca.fit_transform(gfc_norm)\n",
    "components3 = pca3.fit_transform(gfc_norm)\n",
    "\n",
    "print(pandas.DataFrame(components3).head())\n",
    "pca3.explained_variance_ratio_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the most important genes in pc1\n",
    "\n",
    "sparse_comp = pandas.DataFrame(sparse_pca.components_, columns=gfc_norm.columns)\n",
    "print(sparse_comp)\n",
    "pc1 = sparse_comp.loc[0]\n",
    "#pc1_genes = [g.index for g in pc1 if g != 0]\n",
    "print(pc1)\n",
    "pc1 = pc1.apply(abs)\n",
    "#print(pc1.sort_values.__doc__)\n",
    "pc1.sort_values(ascending=False, inplace=True)\n",
    "print(pc1)\n",
    "pc1_gene_list =[gene for gene in pc1.index if pc1[gene] != 0]\n",
    "print(pc1_gene_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some playing around with mutiindex for more useful tables\n",
    "gtm = pandas.read_csv(gtex_tpm_med_fn, sep='\\t', skiprows=2, low_memory=False)\n",
    "gtm.set_index(['Description', 'gene_id'], inplace=True)\n",
    "#tissues = [(col.split(' - ')[0], col.replace(col.split(' - ')[0] + ' - ', '')) for col in gtm.columns]\n",
    "tissues = [col.split(' - ')[0] for col in gtm.columns]\n",
    "subtypes =  [col.replace(col.split(' - ')[0] + ' - ', '') for col in gtm.columns]\n",
    "\n",
    "gtm.columns = [tissues, subtypes]\n",
    "print(gtm.head())\n",
    "\n",
    "print(gtm.loc['TP53', 'Brain'])\n",
    "\n",
    "# Makes some reasonable subtables.  With droplevel seems to work ok."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtm.loc[:, 'Brain']\n",
    "#gtm.loc[:, gtm.columns.get_level_values(1) == 'Cortex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#gtm['skew_Adipose'] = gtm.loc[:, 'Adipose'].apply(lambda row.skew(), axis=1)\n",
    "len(gtm.columns[0])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = gtex_tpm_partial_table_load(gene_list=['TP53'], sample_list=tissue_dict['Lung']['samples'])\n",
    "len(test)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_dict['Brain']['subtypes']['Brain - Hippocampus']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gg = gt[list(gt.columns[:20])]\n",
    "gg.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gmt = gtpm_med.set_index(['Description', 'gene_id']).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gmd = gmt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "low_tissue_expression = gmd.T[gmd.T['75%'] <= 0.01]\n",
    "len(low_tissue_expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# log2 table for measurement\n",
    "g2 = pandas.read_csv('/gsc/resources/expression/GTEx/v7/preprocess/GTEX_log2.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g2.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
