{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Fig3B_finemapping_exp_validation_data.txt',\n",
       " 'Fig3C_CDX8.mainsnp.region.bothtfs.txt',\n",
       " 'Fig3D_CDX8.aFC_EGR1expr.txt',\n",
       " 'Fig5A_enrichment_summary.txt',\n",
       " 'Fig5B_ld_region_count.txt',\n",
       " 'Fig5C_eqtl_gwas_correlation_with_rcp_threshold.txt',\n",
       " 'Fig5D_ukb_rare_variants_coding_model.txt',\n",
       " 'Fig5E_coreg_trait_Pn.txt',\n",
       " 'Fig5E_tissue_sharing_trait_Pn.txt',\n",
       " 'Fig6A_mashr_tissues.txt.gz',\n",
       " 'Fig6B_clustering_similarity_median_rand.txt',\n",
       " 'Fig6C_all_top.z_lfsr.sig.pruned.txt.gz',\n",
       " 'Fig6C_sqtls.z_lfsr.sig.pruned.txt.gz',\n",
       " 'Fig6D_gtex_v8_all_eqtl_categories.txt',\n",
       " 'Fig6_pairwise_roadmap_analysis',\n",
       " 'Fig7B_Whole_Blood_eQTLs.torus_enrichment.txt',\n",
       " 'Fig7B_Whole_Blood_Neutrophil_ieQTLs.torus_enrichment.txt',\n",
       " 'Fig7B_Whole_Blood_Neutrophil_isQTLs.torus_enrichment.txt',\n",
       " 'Fig7B_Whole_Blood_sQTLs.torus_enrichment.txt',\n",
       " 'Fig7C_ieQTL_indep_enrichment_summary.txt',\n",
       " 'Fig7E_Whole_Blood.Neutrophils.eQTL_ieQTL_GWAS_coloc.txt',\n",
       " 'GTEx_Analysis_v8_eQTL.torus_enrichment.txt',\n",
       " 'GTEx_Analysis_v8_eQTL.trans.torus_enrichment.txt',\n",
       " 'GTEx_Analysis_v8_RNAseq_samples.txt',\n",
       " 'GTEx_Analysis_v8_sQTL.torus_enrichment.txt',\n",
       " 'gtex_colors.txt',\n",
       " 'supplement',\n",
       " 'Whole_Blood.v8.independent_eqtls.annotated_Peters_celltype_eQTLs.txt.gz']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# https://github.com/broadinstitute/gtex-v8\n",
    "\n",
    "# The following Python modules are needed to run the notebooks:\n",
    "#  numpy, pandas, scipy, ipython, jupyter, matplotlib, seaborn, qtl\n",
    "\n",
    "# conda config --append channels conda-forge\n",
    "# pip install qtl\n",
    "# conda config --append channels bioconda\n",
    "# conda install pybigwig\n",
    "# conda install bx-python\n",
    "\n",
    "# mkdir pdfs # for plots\n",
    "\n",
    "import os, sys, glob\n",
    "from pprint import pprint\n",
    "\n",
    "# __file__ does not work for python notebooks\n",
    "# python 3.4+ method\n",
    "from pathlib import Path\n",
    "folder = Path().resolve()\n",
    "\n",
    "if os.path.basename(folder) == 'gtex-v8':\n",
    "    data_folder = os.path.join(folder, 'data')\n",
    "elif os.path.basename(folder) == 'data':\n",
    "    data_folder = folder\n",
    "os.listdir(data_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Fig3B_finemapping_exp_validation_data.txt',\n",
      " 'Fig3C_CDX8.mainsnp.region.bothtfs.txt',\n",
      " 'Fig3D_CDX8.aFC_EGR1expr.txt',\n",
      " 'Fig5A_enrichment_summary.txt',\n",
      " 'Fig5B_ld_region_count.txt',\n",
      " 'Fig5C_eqtl_gwas_correlation_with_rcp_threshold.txt',\n",
      " 'Fig5D_ukb_rare_variants_coding_model.txt',\n",
      " 'Fig5E_coreg_trait_Pn.txt',\n",
      " 'Fig5E_tissue_sharing_trait_Pn.txt',\n",
      " 'Fig6A_mashr_tissues.txt.gz',\n",
      " 'Fig6B_clustering_similarity_median_rand.txt',\n",
      " 'Fig6C_all_top.z_lfsr.sig.pruned.txt.gz',\n",
      " 'Fig6C_sqtls.z_lfsr.sig.pruned.txt.gz',\n",
      " 'Fig6D_gtex_v8_all_eqtl_categories.txt',\n",
      " 'Fig6_pairwise_roadmap_analysis',\n",
      " 'Fig7B_Whole_Blood_eQTLs.torus_enrichment.txt',\n",
      " 'Fig7B_Whole_Blood_Neutrophil_ieQTLs.torus_enrichment.txt',\n",
      " 'Fig7B_Whole_Blood_Neutrophil_isQTLs.torus_enrichment.txt',\n",
      " 'Fig7B_Whole_Blood_sQTLs.torus_enrichment.txt',\n",
      " 'Fig7C_ieQTL_indep_enrichment_summary.txt',\n",
      " 'Fig7E_Whole_Blood.Neutrophils.eQTL_ieQTL_GWAS_coloc.txt',\n",
      " 'GTEx_Analysis_v8_eQTL.torus_enrichment.txt',\n",
      " 'GTEx_Analysis_v8_eQTL.trans.torus_enrichment.txt',\n",
      " 'GTEx_Analysis_v8_RNAseq_samples.txt',\n",
      " 'GTEx_Analysis_v8_sQTL.torus_enrichment.txt',\n",
      " 'gtex_colors.txt',\n",
      " 'supplement',\n",
      " 'Whole_Blood.v8.independent_eqtls.annotated_Peters_celltype_eQTLs.txt.gz']\n"
     ]
    }
   ],
   "source": [
    "# Download relevant files\n",
    "\n",
    "os.chdir(data_folder)\n",
    "pprint(os.listdir('.'))\n",
    "org_files = os.listdir('.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:11:36--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:800::2010, 172.217.3.208\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:800::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1562828800 (1.5G) [application/x-tar]\n",
      "Saving to: ‘GTEx_Analysis_v8_eQTL.tar’\n",
      "\n",
      "GTEx_Analysis_v8_eQ 100%[===================>]   1.46G  4.27MB/s    in 6m 14s  \n",
      "\n",
      "2020-02-09 22:17:50 (3.98 MB/s) - ‘GTEx_Analysis_v8_eQTL.tar’ saved [1562828800/1562828800]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:18:27--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_sQTL.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 172.217.14.240\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 569169920 (543M) [application/x-tar]\n",
      "Saving to: ‘GTEx_Analysis_v8_sQTL.tar’\n",
      "\n",
      "GTEx_Analysis_v8_sQ 100%[===================>] 542.80M  4.38MB/s    in 2m 38s  \n",
      "\n",
      "2020-02-09 22:21:05 (3.44 MB/s) - ‘GTEx_Analysis_v8_sQTL.tar’ saved [569169920/569169920]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:21:18--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL_independent.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 172.217.14.240\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 42106880 (40M) [application/x-tar]\n",
      "Saving to: ‘GTEx_Analysis_v8_eQTL_independent.tar’\n",
      "\n",
      "GTEx_Analysis_v8_eQ 100%[===================>]  40.16M  4.14MB/s    in 17s     \n",
      "\n",
      "2020-02-09 22:21:35 (2.40 MB/s) - ‘GTEx_Analysis_v8_eQTL_independent.tar’ saved [42106880/42106880]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:21:37--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_sQTL_independent.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 216.58.193.80\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 15984640 (15M) [application/x-tar]\n",
      "Saving to: ‘GTEx_Analysis_v8_sQTL_independent.tar’\n",
      "\n",
      "GTEx_Analysis_v8_sQ 100%[===================>]  15.24M  2.85MB/s    in 4.8s    \n",
      "\n",
      "2020-02-09 22:21:42 (3.17 MB/s) - ‘GTEx_Analysis_v8_sQTL_independent.tar’ saved [15984640/15984640]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:21:45--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_trans_eGenes_fdr05.txt\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 216.58.193.80\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 22150 (22K) [text/plain]\n",
      "Saving to: ‘GTEx_Analysis_v8_trans_eGenes_fdr05.txt’\n",
      "\n",
      "GTEx_Analysis_v8_tr 100%[===================>]  21.63K  --.-KB/s    in 0.07s   \n",
      "\n",
      "2020-02-09 22:21:46 (312 KB/s) - ‘GTEx_Analysis_v8_trans_eGenes_fdr05.txt’ saved [22150/22150]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:21:47--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_trans_sGenes_fdr05.txt\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 216.58.193.80\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 7332 (7.2K) [text/plain]\n",
      "Saving to: ‘GTEx_Analysis_v8_trans_sGenes_fdr05.txt’\n",
      "\n",
      "GTEx_Analysis_v8_tr 100%[===================>]   7.16K  --.-KB/s    in 0.03s   \n",
      "\n",
      "2020-02-09 22:21:47 (225 KB/s) - ‘GTEx_Analysis_v8_trans_sGenes_fdr05.txt’ saved [7332/7332]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:21:48--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL_expression_matrices.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 216.58.193.80\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1134960640 (1.1G) [application/x-tar]\n",
      "Saving to: ‘GTEx_Analysis_v8_eQTL_expression_matrices.tar’\n",
      "\n",
      "GTEx_Analysis_v8_eQ 100%[===================>]   1.06G  4.34MB/s    in 4m 44s  \n",
      "\n",
      "2020-02-09 22:26:32 (3.81 MB/s) - ‘GTEx_Analysis_v8_eQTL_expression_matrices.tar’ saved [1134960640/1134960640]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:26:55--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL_covariates.tar.gz\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 172.217.14.240\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 6561321 (6.3M) [application/x-tar]\n",
      "Saving to: ‘GTEx_Analysis_v8_eQTL_covariates.tar.gz’\n",
      "\n",
      "GTEx_Analysis_v8_eQ 100%[===================>]   6.26M  2.15MB/s    in 2.9s    \n",
      "\n",
      "2020-02-09 22:26:58 (2.15 MB/s) - ‘GTEx_Analysis_v8_eQTL_covariates.tar.gz’ saved [6561321/6561321]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:27:01--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_sQTL_groups.tar.gz\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 172.217.14.240\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 47041697 (45M) [application/x-tar]\n",
      "Saving to: ‘GTEx_Analysis_v8_sQTL_groups.tar.gz’\n",
      "\n",
      "GTEx_Analysis_v8_sQ 100%[===================>]  44.86M  2.16MB/s    in 15s     \n",
      "\n",
      "2020-02-09 22:27:17 (2.97 MB/s) - ‘GTEx_Analysis_v8_sQTL_groups.tar.gz’ saved [47041697/47041697]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:27:31--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_v8_finemapping_CAVIAR.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:800::2010, 172.217.14.240\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:800::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 901099520 (859M) [application/x-tar]\n",
      "Saving to: ‘GTEx_v8_finemapping_CAVIAR.tar’\n",
      "\n",
      "GTEx_v8_finemapping 100%[===================>] 859.36M  2.09MB/s    in 3m 59s  \n",
      "\n",
      "2020-02-09 22:31:31 (3.59 MB/s) - ‘GTEx_v8_finemapping_CAVIAR.tar’ saved [901099520/901099520]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:31:51--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_v8_finemapping_CaVEMaN.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:800::2010, 172.217.3.176\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:800::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 18196480 (17M) [application/x-tar]\n",
      "Saving to: ‘GTEx_v8_finemapping_CaVEMaN.tar’\n",
      "\n",
      "GTEx_v8_finemapping 100%[===================>]  17.35M  1.53MB/s    in 12s     \n",
      "\n",
      "2020-02-09 22:32:04 (1.40 MB/s) - ‘GTEx_v8_finemapping_CaVEMaN.tar’ saved [18196480/18196480]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:32:04--  https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_v8_finemapping_DAPG.tar\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:800::2010, 172.217.3.176\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:800::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 761794560 (727M) [application/x-tar]\n",
      "Saving to: ‘GTEx_v8_finemapping_DAPG.tar’\n",
      "\n",
      "GTEx_v8_finemapping 100%[===================>] 726.50M  3.08MB/s    in 4m 28s  \n",
      "\n",
      "2020-02-09 22:36:33 (2.71 MB/s) - ‘GTEx_v8_finemapping_DAPG.tar’ saved [761794560/761794560]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:36:42--  https://storage.googleapis.com/gtex_analysis_v8/reference/gencode.v26.GRCh38.genes.gtf\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 172.217.3.208\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 134502408 (128M) [application/octet-stream]\n",
      "Saving to: ‘gencode.v26.GRCh38.genes.gtf’\n",
      "\n",
      "gencode.v26.GRCh38. 100%[===================>] 128.27M  4.68MB/s    in 33s     \n",
      "\n",
      "2020-02-09 22:37:16 (3.85 MB/s) - ‘gencode.v26.GRCh38.genes.gtf’ saved [134502408/134502408]\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:37:16--  https://storage.googleapis.com/gtex_analysis_v8/reference/WGS_Feature_overlap_collapsed_VEP_short_4torus.MAF01.txt.gz\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:809::2010, 172.217.3.208\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:809::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 52580994 (50M) [text/plain]\n",
      "Saving to: ‘WGS_Feature_overlap_collapsed_VEP_short_4torus.MAF01.txt.gz’\n",
      "\n",
      "WGS_Feature_overlap 100%[===================>]  50.14M  4.65MB/s    in 11s     \n",
      "\n",
      "2020-02-09 22:37:28 (4.58 MB/s) - ‘WGS_Feature_overlap_collapsed_VEP_short_4torus.MAF01.txt.gz’ saved [52580994/52580994]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# QTLs\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL.tar && \\\n",
    "    tar xf GTEx_Analysis_v8_eQTL.tar && rm GTEx_Analysis_v8_eQTL.tar\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_sQTL.tar && \\\n",
    "    tar xf GTEx_Analysis_v8_sQTL.tar && rm GTEx_Analysis_v8_sQTL.tar\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL_independent.tar && \\\n",
    "    tar xf GTEx_Analysis_v8_eQTL_independent.tar && rm GTEx_Analysis_v8_eQTL_independent.tar\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_sQTL_independent.tar && \\\n",
    "    tar xf GTEx_Analysis_v8_sQTL_independent.tar && rm GTEx_Analysis_v8_sQTL_independent.tar\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_trans_eGenes_fdr05.txt\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_trans_sGenes_fdr05.txt\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL_expression_matrices.tar && \\\n",
    "    tar xf GTEx_Analysis_v8_eQTL_expression_matrices.tar && rm GTEx_Analysis_v8_eQTL_expression_matrices.tar\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL_covariates.tar.gz && \\\n",
    "    tar xf GTEx_Analysis_v8_eQTL_covariates.tar.gz && rm GTEx_Analysis_v8_eQTL_covariates.tar.gz\n",
    "\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_sQTL_groups.tar.gz && \\\n",
    "    tar xf GTEx_Analysis_v8_sQTL_groups.tar.gz && rm GTEx_Analysis_v8_sQTL_groups.tar.gz\n",
    "\n",
    "# fine mapping results\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_v8_finemapping_CAVIAR.tar && \\\n",
    "    tar xf GTEx_v8_finemapping_CAVIAR.tar && rm GTEx_v8_finemapping_CAVIAR.tar\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_v8_finemapping_CaVEMaN.tar && \\\n",
    "    tar xf GTEx_v8_finemapping_CaVEMaN.tar && rm GTEx_v8_finemapping_CaVEMaN.tar\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_v8_finemapping_DAPG.tar && \\\n",
    "    tar xf GTEx_v8_finemapping_DAPG.tar && rm GTEx_v8_finemapping_DAPG.tar\n",
    "\n",
    "# annotation\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/reference/gencode.v26.GRCh38.genes.gtf\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/reference/WGS_Feature_overlap_collapsed_VEP_short_4torus.MAF01.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:55:41--  https://storage.googleapis.com/gtex_analysis_v8/reference/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:801::2010, 216.58.217.48\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:801::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 404 Not Found\n",
      "2020-02-09 22:55:42 ERROR 404: Not Found.\n",
      "\n",
      "Will not apply HSTS. The HSTS database must be a regular and non-world-writable file.\n",
      "ERROR: could not open HSTS store at '/home/zeno/.wget-hsts'. HSTS will be disabled.\n",
      "--2020-02-09 22:55:42--  https://storage.googleapis.com/gtex_analysis_v8/reference/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.txt.gz\n",
      "Resolving storage.googleapis.com (storage.googleapis.com)... 2607:f8b0:400a:801::2010, 216.58.217.48\n",
      "Connecting to storage.googleapis.com (storage.googleapis.com)|2607:f8b0:400a:801::2010|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 821916072 (784M) [text/plain]\n",
      "Saving to: ‘GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.txt.gz’\n",
      "\n",
      "GTEx_Analysis_2017- 100%[===================>] 783.84M  1.65MB/s    in 3m 55s  \n",
      "\n",
      "2020-02-09 22:59:38 (3.34 MB/s) - ‘GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.txt.gz’ saved [821916072/821916072]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# this is close, but not the original VCF file\n",
    "!wget https://storage.googleapis.com/gtex_analysis_v8/reference/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.txt.gz\n",
    "\n",
    "# this file not exist\n",
    "# !wget https://storage.googleapis.com/gtex_analysis_v8/reference/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz\n",
    "\n",
    "# Need to get it from https://app.terra.bio/ - which has some important rules\n",
    "# * only individual accounts - google logon\n",
    "# * no autologons - no leaving logged in\n",
    "# * Do not access TCGA controlled access data unless dbGaP authorized\n",
    "# * Do not upload personal datasets to the tutorials and do not add worflows (Method Configs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
